#! /usr/bin/env Rscript

library(argparser)
library(data.table)
library(dplyr)
library(tidyr)

p <- arg_parser("Add ENSG ID to all_exons annotation code file generated by leafviz, use new file for GTEx way of mapping clusters to genes")
p <- add_argument(p, "--out", help="New file prefix")
p <- add_argument(p, "--gtf", help="GTF annotation")
argv <- parse_args(p)

gtf <- fread(argv$gtf, data.table = F)
exons <- gtf %>% filter(V3 == "exon") # 1379814 matches leafviz all_exons file
for(i in 1:nrow(exons)){
    exons[i,10] <- strsplit(exons[i,9], split = '["]')[[1]][which(strsplit(exons[i,9], split = '["]')[[1]] == "gene_id ") + 1]
    exons[i,11] <- strsplit(exons[i,9], split = '["]')[[1]][which(strsplit(exons[i,9], split = '["]')[[1]] == "; gene_name ") + 1]
}

exons <- exons %>% separate(V10, c("ensg", "version"), sep = '[.]', remove = FALSE) %>% select(1, 4, 5, 7, ensg, V11)
colnames(exons) <- c("chr", "start", "end", "strand", "gene_id", "gene_name")
write.table(exons, paste0(argv$out,".txt"), col.names = T, row.names = F, quote = F, sep = "\t")