#! /usr/bin/env Rscript

# Refer to GTEX pipeline

library(argparser)
library(data.table)
library(dplyr)
library(tidyr)

p <- arg_parser("Write grouping file and BED file for FastQTL mapping with grouping")
p <- add_argument(p, "--cluster_to_gene", help="Generated by map_clusters_to_genes.R")
p <- add_argument(p, "--bed", help="Original BED file")
p <- add_argument(p, "--info", help="GTF")
p <- add_argument(p, "--grp", help="Grouping file")
p <- add_argument(p, "--bed_out", help="BED output")

argv <- parse_args(p)

# Load data
clu2gene <- fread(argv$cluster_to_gene, data.table = F)
bed <- fread(argv$bed, data.table = F)
info <- fread(argv$info, data.table = F)

# 1. Make grouping file
bed <- bed %>% 
    separate(ID, sep = ':', remove = FALSE, into = c("chr","intron_start","intron_end","clu")) %>%
    unite("cluID", chr, clu, sep = ':')
bed$cluID <- gsub("^","chr",bed$cluID)
bed_info <- bed[,1:7] %>% # 273167      7
    inner_join(clu2gene, by=c("cluID"="clu")) %>% # 263331      8
    select(ID, genes, names)
# one row per intron-gene pair
# if the cluster was mapped to more than one gene, multiple rows for the intron in grouping file
gene_list <- strsplit(bed_info$genes, split = ',')
name_list <- strsplit(bed_info$names, split = ',')
bed_info2 <- data.frame(
    ID = rep(bed_info$ID, sapply(gene_list, length)), 
    gene = unlist(gene_list),
    name = unlist(name_list)) # 311389      3 
bed_info2 <- bed_info2 %>% 
    unite("newID", ID, gene, name, sep = ':', remove = FALSE) %>% 
    select(newID, gene)
write.table(bed_info2, argv$grp, col.names = F, row.names = F, quote = F, sep = "\t")

# 2. Make BED file for grouped mapping
# Add mapped gene in feature ID
# Use gene coordinates
info <- info %>% 
    select(V1, V4, V5, V7, ensg, version, V12) %>%     
    unite("gene", ensg, version, sep = '.')
bed_info2 <- bed_info2 %>% 
    left_join(info, by = "gene") %>% # 311389      7
    separate(newID, into = c("chr","start","end","clu","gene","name"), remove = FALSE, sep = ':') %>%
    unite("intronID", chr, start, end, clu, sep = ':') %>%
    left_join(bed, by = c("intronID" = "ID")) %>%
    select(V1, V4, V5, V7, newID, colnames(bed)[8:(ncol(bed))])

for (i in 1:nrow(bed_info2)){
    if(bed_info2[i,'V7'] == '+'){
        start <- bed_info2[i,'V4']
        bed_info2[i,'V4'] <- start-1
        bed_info2[i,'V5'] <- start}
    if(bed_info2[i,'V7'] == '-'){
        end <- bed_info2[i,'V5']
        bed_info2[i,'V4'] <- end-1
        bed_info2[i,'V5'] <- end}
}

bed_info2$V1 <- gsub("chr", "", bed_info2$V1)
colnames(bed_info2)[1:5] <- c("#Chr", "start", "end", "strand", "ID")
bed_info2 <- bed_info2 %>% select(-strand)
bed_info2 <- bed_info2[order(bed_info2$'#Chr', bed_info2$start),]
write.table(bed_info2, argv$bed_out, col.names = T,row.names = F, quote = F, sep = "\t")