from os.path import join
import os
import numpy as np
import pandas as pd
import sys


configfile: "config.yaml"


"""
rules:
prep
    - expr_prep: expression filter, VST normalization, outlier detection, ComBat, create bed file
    - ancestry_expr: subset ancestries
    - cov: generate covariates files
    - ancestry_cov
nominal pass
    - fastqtl_nominal
    - ancestry_fastqtl_nominal
    - merge_nominal
    - ancestry_merge_nominal
    - call_nominal: identify significant nominal associations
    - ancestry_call_nominal
permutation pass
    - fastqtl_perm
    - ancestry_fastqtl_perm
    - merge_perm
    - ancestry_merge_perm
    - call_perm
    - ancestry_call_perm
GTEX grouped permutation for ALL
    - make_bed_file
    - make_grp_file
    - gtex_fastqtl_grp_perm
    - gtex_grp_perm_write_chunk_log_list
    - gtex_grp_perm_merge_chunk
conditional pass
    - permutations_all_threshold: compute a npval threshold for all expressed genes
    - conditional
    - get_top_variants
susie finemapping
    - vcf_to_dosage
    - make_susie_meta
    - make_susie_expr
    - run_susie
    - merge_susie
    - sort_susie
Torus: functional enrichment
    - (fastqtl_calculate_sebeta)
    - (merge_beta)
    - gtex_fastqtl
    - gtex_write_chunk_log_list
    - gtex_merge_chunk
    - make_torus_input
    - run_torus
"""


rule all:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
            num_hcp=np.arange(10, 101, 10),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
            num_hcp=np.arange(10, 101, 10),
            ancestry=["eur"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
            num_hcp=np.arange(5, 51, 5),
            ancestry=["amr", "afr"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
            ancestry=["eur"],
            num_hcp=60,
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
            ancestry=["amr"],
            num_hcp=15,
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
            ancestry=["afr"],
            num_hcp=20,
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/group.perm.txt.gz",
            num_hcp=70,
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/group.perm.genes.txt.gz",
            num_hcp=70,
        ),
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/conditional_full.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/conditional_top_variants.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm_susie_merged.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm_purity_filtered.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/torus.out",


rule expr_prep:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/tx.counts.scaled.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.tx.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/tx.TPM.tsv",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/geno_subj_{study}.txt",
            study=["walker", "obrien", "werling", "hdbr", "libd"],
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{file}",
            file=[
                "tx.counts.processed.noComBat.tsv",
                "tx.counts.processed.tsv",
                "tx.counts.scaled.normalized.bed.gz",
                "tx.counts.scaled.normalized.bed.gz.tbi",
                "tx.outlier.txt",
            ],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/expr_prep.R"
        ),
        geno_subj_dir="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/",
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/",
        level="tx",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        Rscript {params.script} \
            --gene_count {input[0]} \
            --gene_gtf {input[1]} \
            --gene_tpm {input[2]} \
            --geno_subj_dir {params.geno_subj_dir} \
            --outdir {params.outdir} \
            --level {params.level}

        bgzip {params.outdir}{params.level}.counts.scaled.normalized.bed
        tabix -p bed {params.outdir}{params.level}.counts.scaled.normalized.bed.gz
        """


rule ancestry_expr:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{file}",
            file=[
                "tx.counts.processed.noComBat.tsv",
                "tx.counts.processed.tsv",
                "tx.counts.scaled.normalized.bed.gz",
                "tx.counts.scaled.normalized.bed.gz.tbi",
            ],
        ),
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/ancestry_list/list.{ancestry}.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{{ancestry}}/{file}",
            file=[
                "tx.counts.processed.noComBat.tsv",
                "tx.counts.processed.tsv",
                "tx.counts.scaled.normalized.bed.gz",
                "tx.counts.scaled.normalized.bed.gz.tbi",
            ],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/ancestry_expr.R",
        data_dir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/",
        level="tx",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        mkdir -p {params.data_dir}{wildcards.ancestry}
        Rscript {params.script} \
            --data_dir {params.data_dir} \
            --ancestry {wildcards.ancestry} \
            --subj_list {input[4]} \
            --level {params.level}
        bgzip {params.data_dir}{wildcards.ancestry}/{params.level}.counts.scaled.normalized.bed
        tabix -p bed {params.data_dir}{wildcards.ancestry}/{params.level}.counts.scaled.normalized.bed.gz
        """


rule cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.processed.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{num_hcp}hcp_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


rule ancestry_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/tx.counts.processed.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/ancestry_cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


rule fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule ancestry_fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{{num_hcp}}hcp/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=160,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/",
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_nominal.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


rule ancestry_merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=160,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/",
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_nominal.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


rule fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/70hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=12,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 200 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule ancestry_fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=12,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 200 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_perm:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 201, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 201, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/",
        script=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_perm.R"
        ),
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


rule ancestry_merge_perm:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 201, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 201, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/{ancestry}_perm_{num_hcp}HCP/",
        script=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_perm.R"
        ),
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


############################## GTEx way, grouped permutations ##############################
rule make_bed_file:
    input:
        tx2gene="/u/project/gandalm/cindywen/isoform_twas/salmon/tx2gene_gencode_v33_noGeneVersion.tsv",
        bed="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        info=(
            "/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.gene.noVersion.tsv"
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.bed.gz.tbi",
    resources:
        mem_gb=4,
        time_min=30,
    params:
        bed_out="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.bed",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9
        Rscript iso_make_bed_file.R \
            --tx2gene {input.tx2gene} \
            --bed {input.bed} \
            --info {input.info} \
            --bed_out {params.bed_out}
        bgzip {params.bed_out}
        tabix -p bed {params.bed_out}.gz
        """


rule make_grp_file:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/salmon/tx2gene_gencode_v33_noGeneVersion.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.txt",
    resources:
        num_gb=4,
        time_min=30,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript iso_make_grp_file.R \
            --tx2gene {input[0]} \
            --out {output[0]}
        """


# The following three rules are basically running run_FastQTL_threaded.py in GTEx's version of FastQTL
# Celine ran this rule in bash script, output files won't be removed if error when no cis variants in chunk
rule gtex_fastqtl_grp_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
        grp_file="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx_grp.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/chunk{chunk}.log",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=240,
        num_cores=10,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/",
    shell:
        """
        mkdir -p {params.out_dir}
        . /u/local/Modules/default/init/modules.sh
        module load gcc/9.3.0
        /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/bin/fastQTL.static \
            --vcf {input.geno} \
            --bed {input.expr} \
            --window 1e6 \
            --cov {input.cov} \
            --grp {input.grp_file} \
            --permute 1000 10000 \
            --seed 123 \
            --exclude-samples {input.rel_file} \
            --chunk {wildcards.chunk} 500 \
            --out {output[0]} \
            --log {output[1]} \
            --normal
        touch {output[2]}
        """


rule gtex_grp_perm_write_chunk_log_list:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{{num_hcp}}hcp/chunk{chunk}.{file}",
            file=["txt.gz", "log", "txt.gz.done"],
            chunk=np.arange(1, 501, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/list_log.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        for i in {{1..500..1}}; do
        echo /u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{wildcards.num_hcp}hcp/chunk${{i}}.txt.gz >> {output[0]}
        echo /u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{wildcards.num_hcp}hcp/chunk${{i}}.log >> {output[1]}
        done
        """


rule gtex_grp_perm_merge_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/list_log.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/group.perm.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/group.perm.genes.txt.gz",
    resources:
        num_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_grp_perm_{num_hcp}hcp/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        module load R/3.6.0
        python3 /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/python/merge_chunks.py \
            {input[0]} \
            {input[1]} \
            group.perm \
            --fdr 0.05 \
            -o {params.outdir} \
            --permute
        """


############################## QTLtools conditional pass ##############################
rule permutations_all_threshold:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/all.chunks.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/permutations_all_threshold.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/permutations_all_threshold.R",
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}permutations_all_threshold.txt
        """


rule make_qtltools_bed:
    input:
        fastqtl_expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/qtltools.tx.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/qtltools.tx.counts.scaled.normalized.bed.gz.tbi",
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        zcat {input.fastqtl_expr} | awk '{{ $4=$4" . +"; print $0 }}' | tr " " "\t" | bgzip -c > {output[0]}
        tabix -p bed {output[0]}
        """


rule conditional:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/qtltools.tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/70hcp_cov.txt",
        threshold="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/permutations_all_threshold.txt.gz",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/chunk{chunk}.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/chunk{chunk}.txt.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/",
    shell:
        """
        mkdir -p {params.outdir}
        {config[QTLTOOLS]} cis \
            --vcf {input.geno} \
            --bed {input.expr} \
            --cov {input.cov} \
            --mapping {input.threshold} \
            --chunk {wildcards.chunk} 200 \
            --out {output[0]} \
            --exclude-samples {input.rel_file} \
            --seed 123 \
            --window 1000000 \
            --normal
        touch {output[1]}
        """


rule get_top_variants:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/chunk{chunk}.txt.done",
            chunk=np.arange(1, 201, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/conditional_full.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/conditional_top_variants.txt",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_conditional_70hcp/",
    shell:
        """
        cat {params.outdir}chunk*.txt > {params.outdir}conditional_full.txt
        cat {params.outdir}conditional_full.txt | awk '{{ if ($19 == 1) print $0}}' > {params.outdir}conditional_top_variants.txt
        """


################################### susie finemapping ###################################
rule vcf_to_dosage:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{file}",
            file=[
                "4_columns_iso.tsv",
                "sample_list_iso.tsv",
                "header_iso.tsv",
                "header_row_iso.tsv.gz",
                "dose_matrix_iso.tsv.gz",
                "filtered.hg19.sorted.removeTxOutlier.dose.tsv.gz",
                "filtered.hg19.sorted.removeTxOutlier.dose.tsv.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load htslib/1.9

        # Extract header
        printf 'CHROM\nPOS\nREF\nALT\n' > {output[0]}
        bcftools query -l {input.vcf} > {output[1]}
        cat {output[0]} {output[1]} > {output[2]}
        {config[CSVTK]} transpose {output[2]} -T | gzip > {output[3]}

        # Extract dosage and merge
        bcftools +dosage {input.vcf} -- -t GT | tail -n+2 | gzip > {output[4]}
        zcat {output[3]} {output[4]} | bgzip > {params.prefix}.dose.tsv.gz
        tabix -s1 -b2 -e2 -S1 {params.prefix}.dose.tsv.gz
        """


rule make_susie_meta:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/phenotype_meta.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/sample_meta.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_meta.R",
        qtl_group="mixed_cisisoqtl_70hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --bed {input[0]} \
            --qtl_group {params.qtl_group} \
            --exclude_rel {input[1]}
        """


rule make_susie_expr:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.processed.susie.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_expr.R",
        outname="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.processed.susie.tsv",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --input {input[0]} \
            --outname {params.outname} \
            --exclude_rel {input[1]}
        """


rule run_susie:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.processed.susie.tsv",
        pheno_meta="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/phenotype_meta.tsv",
        sample_meta=(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/sample_meta.tsv"
        ),
        sig_pheno="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/sig_pheno.txt",
        genotype_matrix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.dose.tsv.gz",
        covariates=(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/70hcp_cov_626.txt"
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm.{{batch}}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
        ),
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=120,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/run_susie_customized.R",
        cisdistance=1000000,
        permuted="true",
        qtl_group="mixed_cisisoqtl_70hcp_perm",
        out_dir=(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/"
        ),
        quant_method="gene_counts",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        mkdir -p {params.out_dir}

        Rscript {params.script} \
            --expression_matrix {input.expr} \
            --phenotype_meta {input.pheno_meta} \
            --sample_meta {input.sample_meta} \
            --sig_pheno {input.sig_pheno} \
            --covariates {input.covariates} \
            --genotype_matrix {input.genotype_matrix} \
            --chunk '{wildcards.batch} {config[SUSIE_N_BATCH]}' \
            --cisdistance {params.cisdistance} \
            --out_prefix {params.out_dir}{params.qtl_group}.{wildcards.batch}_{config[SUSIE_N_BATCH]} \
            --permuted {params.permuted} \
            --qtl_group {params.qtl_group} \
            --quant_method {params.quant_method}
        """


rule merge_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm.{batch}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
            batch=np.arange(1, 501, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.txt | bgzip -c > {params.prefix}.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.cred.txt | bgzip -c > {params.prefix}.cred.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.snp.txt | bgzip -c > {params.prefix}.snp.txt.gz
        """


rule sort_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm_susie_merged.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm_purity_filtered.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/susie_finemap/mixed_cisisoqtl_70hcp_perm",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        gunzip -c {input[0]} > {params.prefix}_susie_merged.txt
        (head -n 1 {params.prefix}_susie_merged.txt && tail -n +2 {params.prefix}_susie_merged.txt | sort -k3 -k4n ) | bgzip > {params.prefix}_purity_filtered.txt.gz
        """


################################### Torus functional enrichment ###################################
# rule fastqtl_calculate_sebeta:
#     input:
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/chunk{chunk}.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/significant_assoc.txt",
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_chunk{chunk}.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_chunk{chunk}.txt.gz.done",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_chunk{chunk}",
#         # num_comparisons=748456165
#     shell:
#         """
#         . /u/local/Modules/default/init/modules.sh
#         module load R/3.6.0
#         Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/calculate_sebeta.R \
#             --input {input[0]} \
#             --output {params.prefix}.txt \
#             --sig {input[1]}
#         gzip {params.prefix}.txt
#         touch {params.prefix}.txt.gz.done
#         """


# rule merge_sebeta:
#     input:
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_chunk{chunk}.txt.gz",
#             chunk=np.arange(1, 101, 1),
#         ),
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_chunk{chunk}.txt.gz.done",
#             chunk=np.arange(1, 101, 1),
#         ),
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_all_chunks.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_all_chunks.txt.gz.done",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/sebeta_significant_",
#     shell:
#         """
#         zcat {params.prefix}chunk*.txt.gz | gzip -c > {params.prefix}all_chunks.txt.gz
#         touch {params.prefix}all_chunks.txt.gz.done
#         """


rule gtex_fastqtl:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/tx.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeTxOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.log",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp",
    shell:
        """
        # mkdir -p {params.outdir}
        . /u/local/Modules/default/init/modules.sh
        module load gcc/9.3.0
        /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/bin/fastQTL.static \
            --vcf {input.geno} \
            --bed {input.expr} \
            --window 1e6 \
            --cov {input.cov} \
            --seed 123 \
            --exclude-samples {input.rel_file} \
            --chunk {wildcards.chunk} 100 \
            --out {output[0]} \
            --log {output[1]} \
            --normal
        touch {output[0]}.done
        """


rule gtex_write_chunk_log_list:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{{num_hcp}}hcp/gtex_chunk{chunk}.{file}",
            file=["txt.gz", "log", "txt.gz.done"],
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/list_log.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        for i in {{1..100..1}}; do
        echo /u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp/gtex_chunk${{i}}.txt.gz >> {output[0]}
        echo /u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp/gtex_chunk${{i}}.log >> {output[1]}
        done
        """


rule gtex_merge_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/list_log.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex.allpairs.txt.gz",
    resources:
        num_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_{num_hcp}hcp/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        module load R/3.6.0
        python3 /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/python/merge_chunks.py \
            {input[0]} \
            {input[1]} \
            gtex \
            --fdr 0.05 \
            -o {params.outdir} \
        """


rule make_torus_input:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/gtex.allpairs.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/gtex.allpairs.torus.txt.gz",
    resources:
        mem_gb=4,
        time_min=120,
    shell:
        """
        zcat {input[0]} | tail -n+2 | awk '{{print $1,$2,$3,$7,$8,$9}}' OFS=' ' | gzip > {output[0]}
        """


# Celine ran this in bash
rule run_torus:
    input:
        annot="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep.txt.gz",
        fastqtl="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/gtex.allpairs.torus.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_nominal_70hcp/torus.out",
    resources:
        mem_gb=8,
        time_min=4320,
        num_cores=12,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.3
        /u/project/gandalm/shared/apps/torus-master/src/torus \
            -d {input.fastqtl} \
            --fastqtl \
            -est \
            -annot {input.annot} \
            > {output[0]}
        """
