from os.path import join
import os
import numpy as np
import pandas as pd
import sys


configfile: "config.yaml"


"""
rules:
- make_coord_file
- make_set_file_mixed_top_eqtl
- make_annot_mixed_top_eqtl
- partition_h2_mixed_top_eqtl
- make_set_file_mixed_top_isoqtl
- make_annot_mixed_top_isoqtl
- partition_h2_mixed_top_isoqtl
"""

GWAS_DIC = {s["trait"]: s["file"] for s in config["GWAS_LIST"]}
WINDOW = [0, 250, 500, 1000, 2500]


rule all:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/sig_pheno_gene.txt",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{GWAS_trait}_mixed_top_eqtl_{window}bp_window.{file}",
            window=WINDOW,
            file=["cov", "delete", "log", "part_delete", "results"],
            GWAS_trait=list(GWAS_DIC.keys()),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{GWAS_trait}_mixed_top_isoqtl_{window}bp_window.{file}",
            window=WINDOW,
            file=["cov", "delete", "log", "part_delete", "results"],
            GWAS_trait=list(GWAS_DIC.keys()),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{GWAS_trait}_mixed_susie_{window}bp_window.{file}",
            window=WINDOW,
            file=["cov", "delete", "log", "part_delete", "results"],
            GWAS_trait=list(GWAS_DIC.keys()),
        ),


rule make_coord_file:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.vcf.gz",
    output:
        info=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_info.tsv"
        ),
        coord="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_coord.tsv",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        script="scripts/variant_coord.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load R/3.6.0

        bcftools query -f '%CHROM %POS %ID\n' {input.vcf} -o {output.info}
        Rscript {params.script} \
            --info {output.info} \
            --out {output.coord}
        """


rule make_set_file_mixed_top_eqtl:
    input:
        input_file="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/sig_pheno.txt",
    output:
        set_file=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_top_eqtl.txt"
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/set_file.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --input {input.input_file} \
            --out {output.set_file}
        """


rule make_annot_mixed_top_eqtl:
    input:
        set_file=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_top_eqtl.txt"
        ),
        coord_file="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_coord.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_top_eqtl_{{window}}bp_window.{{chr_num}}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_top_eqtl_{window}bp_window",
        script="scripts/make_annot_customed.py",
    shell:
        """
        mkdir -p {params.annot_dir}
        # I had problem with /u/home/c/cindywen/anaconda3/etc/profile.d/conda.sh: line 55: PS1: unbound variable. Found this workaround:
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "Make ldsc-friendly annotation files"
        {params.script} \
            --gene-set-file {input.set_file} \
            --gene-coord-file {input.coord_file} \
            --bimfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num}.bim \
            --annot-file {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --windowsize {wildcards.window}

        echo "Computing LD scores with the annot file {{params.annot_prefix}}.{{wildcards.chr_num}}.annot.gz"
        {config[LDSC_PATH]}ldsc.py \
            --l2 \
            --bfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num} \
            --print-snps {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/print_snps.txt \
            --ld-wind-cm 1 \
            --annot {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --thin-annot \
            --out {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}

        set -eu
        """


rule partition_h2_mixed_top_eqtl:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_top_eqtl_{{window}}bp_window.{chr_num}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
            chr_num=np.arange(1, 23, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{{GWAS_trait}}_mixed_top_eqtl_{{window}}bp_window.{file}",
            file=["cov", "delete", "log", "part_delete", "results"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/",
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_top_eqtl_{window}bp_window",
        GWAS_file=lambda wildcards: GWAS_DIC[wildcards.GWAS_trait],
    shell:
        """
        mkdir -p {params.out_dir}
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "GWAS trait: {wildcards.GWAS_trait}"
        {config[LDSC_PATH]}ldsc.py \
            --h2 {params.GWAS_file}.sumstats.gz \
            --ref-ld-chr {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/baseline.,{params.annot_dir}{params.annot_prefix}. \
            --frqfile-chr {config[LDSC_PATH]}LDSCORE/1000G_Phase3_frq/1000G.EUR.QC. \
            --w-ld-chr {config[LDSC_PATH]}LDSCORE/weights_hm3_no_hla/weights. \
            --overlap-annot \
            --print-cov \
            --print-coefficients \
            --print-delete-vals \
            --out {params.out_dir}{wildcards.GWAS_trait}_{params.annot_prefix}

        set -eu
        """


rule make_set_file_mixed_top_isoqtl:
    input:
        input_file="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/sig_pheno.txt",
        tx2gene="/u/project/gandalm/cindywen/isoform_twas/salmon/tx2gene_gencode_v33_noGeneVersion.tsv",
    output:
        set_file=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_top_isoqtl.txt"
        ),
        out2="/u/project/gandalm/cindywen/isoform_twas/isoqtl_new/results/mixed_perm_70hcp/sig_pheno_gene.txt",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/set_file_isoqtl.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --input {input.input_file} \
            --out {output.set_file} \
            --tx2gene {input.tx2gene} \
            --out2 {output.out2}
        """


rule make_annot_mixed_top_isoqtl:
    input:
        set_file=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_top_isoqtl.txt"
        ),
        coord_file="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_coord.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_top_isoqtl_{{window}}bp_window.{{chr_num}}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_top_isoqtl_{window}bp_window",
        script="scripts/make_annot_customed.py",
    shell:
        """
        mkdir -p {params.annot_dir}
        # I had problem with /u/home/c/cindywen/anaconda3/etc/profile.d/conda.sh: line 55: PS1: unbound variable. Found this workaround:
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "Make ldsc-friendly annotation files"
        {params.script} \
            --gene-set-file {input.set_file} \
            --gene-coord-file {input.coord_file} \
            --bimfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num}.bim \
            --annot-file {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --windowsize {wildcards.window}

        echo "Computing LD scores with the annot file {{params.annot_prefix}}.{{wildcards.chr_num}}.annot.gz"
        {config[LDSC_PATH]}ldsc.py \
            --l2 \
            --bfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num} \
            --print-snps {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/print_snps.txt \
            --ld-wind-cm 1 \
            --annot {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --thin-annot \
            --out {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}

        set -eu
        """


rule partition_h2_mixed_top_isoqtl:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_top_isoqtl_{{window}}bp_window.{chr_num}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
            chr_num=np.arange(1, 23, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{{GWAS_trait}}_mixed_top_isoqtl_{{window}}bp_window.{file}",
            file=["cov", "delete", "log", "part_delete", "results"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/",
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_top_isoqtl_{window}bp_window",
        GWAS_file=lambda wildcards: GWAS_DIC[wildcards.GWAS_trait],
    shell:
        """
        mkdir -p {params.out_dir}
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "GWAS trait: {wildcards.GWAS_trait}"
        {config[LDSC_PATH]}ldsc.py \
            --h2 {params.GWAS_file}.sumstats.gz \
            --ref-ld-chr {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/baseline.,{params.annot_dir}{params.annot_prefix}. \
            --frqfile-chr {config[LDSC_PATH]}LDSCORE/1000G_Phase3_frq/1000G.EUR.QC. \
            --w-ld-chr {config[LDSC_PATH]}LDSCORE/weights_hm3_no_hla/weights. \
            --overlap-annot \
            --print-cov \
            --print-coefficients \
            --print-delete-vals \
            --out {params.out_dir}{wildcards.GWAS_trait}_{params.annot_prefix}

        set -eu
        """


rule make_set_file_mixed_susie:
    input:
        susie="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm_purity_filtered.txt.gz",
        info=(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_info.tsv"
        ),
    output:
        set_file="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_susie_var_in_cs.txt",
    resources:
        mem_gb=6,
        time_min=60,
    params:
        script="scripts/set_file_susie.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --susie {input.susie} \
            --info {input.info} \
            --out {output.set_file}
        """


rule make_annot_mixed_susie:
    input:
        set_file="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_susie_var_in_cs.txt",
        coord_file="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_coord.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_susie_{{window}}bp_window.{{chr_num}}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_susie_{window}bp_window",
        script="scripts/make_annot_customed.py",
    shell:
        """
        mkdir -p {params.annot_dir}
        # I had problem with /u/home/c/cindywen/anaconda3/etc/profile.d/conda.sh: line 55: PS1: unbound variable. Found this workaround:
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "Make ldsc-friendly annotation files"
        {params.script} \
            --gene-set-file {input.set_file} \
            --gene-coord-file {input.coord_file} \
            --bimfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num}.bim \
            --annot-file {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --windowsize {wildcards.window}

        echo "Computing LD scores with the annot file {{params.annot_prefix}}.{{wildcards.chr_num}}.annot.gz"
        {config[LDSC_PATH]}ldsc.py \
            --l2 \
            --bfile {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_plink/1000G.EUR.QC.{wildcards.chr_num} \
            --print-snps {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/print_snps.txt \
            --ld-wind-cm 1 \
            --annot {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}.annot.gz \
            --thin-annot \
            --out {params.annot_dir}{params.annot_prefix}.{wildcards.chr_num}

        set -eu
        """


rule partition_h2_mixed_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/mixed_susie_{{window}}bp_window.{chr_num}.{file}",
            file=["annot.gz", "l2.ldscore.gz", "l2.M", "l2.M_5_50", "log"],
            chr_num=np.arange(1, 23, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/{{GWAS_trait}}_mixed_susie_{{window}}bp_window.{file}",
            file=["cov", "delete", "log", "part_delete", "results"],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/results/",
        annot_dir="/u/project/gandalm/cindywen/isoform_twas/sLDSC/annot/",
        annot_prefix="mixed_susie_{window}bp_window",
        GWAS_file=lambda wildcards: GWAS_DIC[wildcards.GWAS_trait],
    shell:
        """
        mkdir -p {params.out_dir}
        set +eu 
        source ~/anaconda3/bin/activate ldsc

        echo "GWAS trait: {wildcards.GWAS_trait}"
        {config[LDSC_PATH]}ldsc.py \
            --h2 {params.GWAS_file}.sumstats.gz \
            --ref-ld-chr {config[LDSC_PATH]}LDSCORE/1000G_EUR_Phase3_baseline/baseline.,{params.annot_dir}{params.annot_prefix}. \
            --frqfile-chr {config[LDSC_PATH]}LDSCORE/1000G_Phase3_frq/1000G.EUR.QC. \
            --w-ld-chr {config[LDSC_PATH]}LDSCORE/weights_hm3_no_hla/weights. \
            --overlap-annot \
            --print-cov \
            --print-coefficients \
            --print-delete-vals \
            --out {params.out_dir}{wildcards.GWAS_trait}_{params.annot_prefix}

        set -eu
        """
