from os.path import join
import os
import numpy as np
import pandas as pd
import sys

configfile: "config.yaml"

"""
rules:
    expr_prep: expression filter, VST normalization, outlier detection, ComBat, create bed file
    ancestry_expr: subset ancestries
    cov: generate covariates files
    ancestry_cov:
    fastqtl_nominal:
    ancestry_fastqtl_nominal:
    call_nominal: identify significant nominal associations
    ancestry_call_nominal:
    fastqtl_perm:
    ancestry_fastqtl_perm:
    call_perm:
    ancestry_call_perm:
"""
rule all:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/{file}", file=["all.chunks.txt.gz", "significant_assoc.txt", "significant_feature_count.txt"], num_hcp=np.arange(10, 101, 10)),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}", file=["all.chunks.txt.gz", "significant_assoc.txt", "significant_feature_count.txt"], num_hcp=np.arange(10, 101, 10), ancestry=["eur"]),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}", file=["all.chunks.txt.gz", "significant_assoc.txt", "significant_feature_count.txt"], num_hcp=np.arange(5, 51, 5), ancestry=["amr", "afr"]),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/{file}", file=["all.chunks.txt.gz", "sig_pheno.txt"]),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}", ancestry=["eur"], num_hcp=50, file=["all.chunks.txt.gz", "sig_pheno.txt"]),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}", ancestry=["amr"], num_hcp=15, file=["all.chunks.txt.gz", "sig_pheno.txt"]),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}", ancestry=["afr"], num_hcp=20, file=["all.chunks.txt.gz", "sig_pheno.txt"])



rule expr_prep:
    input:  
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.scaled.counts.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.gene.noVersion.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.TPM.tsv",
        expand("/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/geno_subj_{study}.txt", study=["walker","obrien","werling","hdbr","libd"])
    output: 
    	expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{file}", file=["gene.counts.processed.noComBat.tsv", "gene.counts.processed.tsv","gene.counts.scaled.normalized.bed.gz","gene.counts.scaled.normalized.bed.gz.tbi", "gene.outlier.txt"])
    resources:  
        mem_gb=8,
        time_min=60
    params: 
        script="scripts/expr_prep.R",
        geno_subj_dir="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/",
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9
        
        Rscript {params.script} \
            --gene_count {input[0]} \
            --gene_gtf {input[1]} \
            --gene_tpm {input[2]} \
            --geno_subj_dir {params.geno_subj_dir} \
            --outdir {params.outdir}
        
        bgzip {params.outdir}gene.counts.scaled.normalized.bed
        tabix -p bed {params.outdir}gene.counts.scaled.normalized.bed.gz
        """

rule ancestry_expr:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{file}", file=["gene.counts.processed.noComBat.tsv", "gene.counts.processed.tsv","gene.counts.scaled.normalized.bed.gz","gene.counts.scaled.normalized.bed.gz.tbi"]),
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/ancestry_list/list.{ancestry}.tsv"
    output:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{{ancestry}}/{file}", file=["gene.counts.processed.noComBat.tsv", "gene.counts.processed.tsv","gene.counts.scaled.normalized.bed.gz","gene.counts.scaled.normalized.bed.gz.tbi"])
    resources:
        mem_gb=4,
        time_min=60
    params:
        script="scripts/ancestry_expr.R",
        data_dir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        mkdir -p {params.data_dir}{wildcards.ancestry}
        Rscript {params.script} \
            --data_dir {params.data_dir} \
            --ancestry {wildcards.ancestry} \
            --subj_list {input[4]}
        bgzip {params.data_dir}{wildcards.ancestry}/gene.counts.scaled.normalized.bed
        tabix -p bed {params.data_dir}{wildcards.ancestry}/gene.counts.scaled.normalized.bed.gz
        """

rule cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.tsv",
        picard="/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv",
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{num_hcp}hcp_cov.txt"
    resources:
        mem_gb=8,
        time_min=60
    params:
        script="scripts/cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """

# I had to change output file name to make it unambiguous with rule cov
rule ancestry_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.tsv",
        picard="/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv",
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt"
    resources:
        mem_gb=8,
        time_min=60
    params:
        script="scripts/ancestry_cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """ 

rule fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{num_hcp}hcp_cov.txt"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz.done"
    resources:
        mem_gb=4,
        time_min=120
    params: 
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp"
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal
        touch {output[0]}.done
        """

rule ancestry_fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz.done"
    resources:
        mem_gb=16,
        time_min=120
    params: 
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP"
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal
        touch {output[0]}.done
        """

rule call_nominal:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz", chunk=np.arange(1, 101, 1)),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz.done", chunk=np.arange(1, 101, 1))
    output:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/{file}", file=["all.chunks.txt.gz", "significant_assoc.txt", "significant_feature_count.txt"])
    resources:
        mem_gb=24,
        time_min=160
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/",
        script="scripts/call_nominal.R"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}/all.chunks.txt.gz
        Rscript {params.script} \
            --outdir {params.outdir}
        """

rule ancestry_call_nominal:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz", chunk=np.arange(1, 101, 1)),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done", chunk=np.arange(1, 101, 1))
    output:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/{file}", file=["all.chunks.txt.gz", "significant_assoc.txt", "significant_feature_count.txt"])
    resources:
        mem_gb=24,
        time_min=160
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/",
        script="scripts/call_nominal.R"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}/all.chunks.txt.gz
        Rscript {params.script} \
            --outdir {params.outdir}
        """

# Had trouble with 500 chunks, some chunks do not have any variants
rule fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/90hcp_cov.txt"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz",
	"/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz.done"
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=4
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp"
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 300 \
                          --seed 123 \
                          --window 1e6 \
                          --normal
        touch {output[0]}.done
        """

rule ancestry_fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt"
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz",
	"/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz.done"
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP"
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 300 \
                          --seed 123 \
                          --window 1e6 \
                          --normal
        touch {output[0]}.done
        """

rule call_perm:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz", chunk=np.arange(1,301,1)),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz.done", chunk=np.arange(1,301,1))
    output:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/{file}", file=["all.chunks.txt.gz", "sig_pheno.txt"])
    resources:
        mem_gb=6,
        time_min=120
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/",
        script="scripts/call_perm.R"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        Rscript {params.script} \
            --outdir {params.outdir}
        """

rule ancestry_call_perm:
    input:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz", chunk=np.arange(1,301,1)),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done", chunk=np.arange(1,301,1))
    output:
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/{file}", file=["all.chunks.txt.gz", "sig_pheno.txt"])
    resources:
        mem_gb=6,
        time_min=120
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/",
        script="scripts/call_perm.R"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        Rscript {params.script} \
            --outdir {params.outdir}
        """
