from os.path import join
import os
import numpy as np
import pandas as pd
import sys


configfile: "config.yaml"


"""
rules:
prep
    - expr_prep: expression filter, VST normalization, outlier detection, ComBat, create bed file
    - ancestry_expr: subset ancestries
    - cov: generate covariates files
    - ancestry_cov
nominal pass
    - fastqtl_nominal
    - ancestry_fastqtl_nominal
    - merge_nominal
    - ancestry_merge_nominal
    - call_nominal: identify significant nominal associations
    - ancestry_call_nominal
permutation pass
    - fastqtl_perm
    - ancestry_fastqtl_perm
    - merge_perm
    - ancestry_merge_perm
    - call_perm
    - ancestry_call_perm
conditional pass
    - permutations_all_threshold: compute a npval threshold for all expressed genes
    - conditional
    - get_top_variants
susie finemapping
    - vcf_to_dosage
    - make_susie_meta
    - make_susie_expr
    - run_susie
    - merge_susie
    - sort_susie
susie finemapping ancsetry
    (as above)
    - locuszoom: also see susie_analysis.ipynb 
Ancestry eQTL effect size
    - make_effect_size_scatter_eur_amr, eur_afr, afr_amr: all nominal associations
    - scatter_eur_afr, eur_amr, afr_amr: x-axis pop1 eGene-eQTL, y-axis sig or non-sig
Torus: functional enrichment
    - make_annot_ensembl: use ensembl regulatory build to annotate our variants
    - run_vep: ensembl variant effect predictor
    - select_consequences: select rsID and consequence/annotation columns
    - add_vep: add VEP to annotation
    - merge_annot_ensembl_vep: merge chr
    - (fastqtl_calculate_sebeta): too slow for all nominal association, run GTEx's FastQTL to get sebeta
    - (merge_sebeta)
    - gtex_fastqtl
    - 
    - run_torus
PAINTOR: multi-ethnic fine-mapping
    - make_eur_coord: as in sLDSC for ALL, make variant coord file for EUR. ALL does not cover all shared variants between EUR, AMR, AFR
    - make_eur_annot: need variant annot for shared variants between EUR, AMR, AFR
    - merge_eur_annot:
"""


rule all:
    input:
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/{file}",
        #     file=[
        #         "all_assoc.txt.gz",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        #     num_hcp=np.arange(10, 101, 10),
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}",
        #     file=[
        #         "all_assoc.txt.gz",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        #     num_hcp=np.arange(10, 101, 10),
        #     ancestry=["eur"],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/{file}",
        #     file=[
        #         "all_assoc.txt.gz",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        #     num_hcp=np.arange(5, 51, 5),
        #     ancestry=["amr", "afr"],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry=["eur"],
        #     num_hcp=50,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry=["amr"],
        #     num_hcp=15,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry=["afr"],
        #     num_hcp=25,
        # ),
        # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm_purity_filtered.txt.gz",
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{ancestry}_ciseqtl_{num_hcp}hcp_perm_purity_filtered.txt.gz",
        #     zip,
        #     ancestry=["eur", "amr", "afr"],
        #     num_hcp=[50, 15, 25],
        # ),
        expand("/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/locuszoom_{ancestry}_{num_hcp}HCP_chr15_81279136.pdf", zip, ancestry=["eur", "amr", "afr"],num_hcp=[50, 15, 25]),
        # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/conditional_full.txt",
        # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_eur_amr.png",
        # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_eur_afr.png",
        # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_afr_amr.png",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/perm_scatter_eur_afr.png",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/perm_scatter_eur_amr.png",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/torus.out",
        # "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot.txt.gz",


rule expr_prep:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.scaled.counts.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.gene.noVersion.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/salmon/expression.final/gene.noVersion.TPM.tsv",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/geno_subj_{study}.txt",
            study=["walker", "obrien", "werling", "hdbr", "libd"],
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{file}",
            file=[
                "gene.counts.processed.noComBat.tsv",
                "gene.counts.processed.tsv",
                "gene.counts.scaled.normalized.bed.gz",
                "gene.counts.scaled.normalized.bed.gz.tbi",
                "gene.outlier.txt",
            ],
        ),
    resources:
        mem_gb=8,
        time_min=60,
    params:
        script="scripts/expr_prep.R",
        geno_subj_dir="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/",
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/",
        level="gene",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        Rscript {params.script} \
            --gene_count {input[0]} \
            --gene_gtf {input[1]} \
            --gene_tpm {input[2]} \
            --geno_subj_dir {params.geno_subj_dir} \
            --outdir {params.outdir} \
            --level {params.level}

        bgzip {params.outdir}{params.level}.counts.scaled.normalized.bed
        tabix -p bed {params.outdir}{params.level}.counts.scaled.normalized.bed.gz
        """


rule ancestry_expr:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{file}",
            file=[
                "gene.counts.processed.noComBat.tsv",
                "gene.counts.processed.tsv",
                "gene.counts.scaled.normalized.bed.gz",
                "gene.counts.scaled.normalized.bed.gz.tbi",
            ],
        ),
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/ancestry_list/list.{ancestry}.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{{ancestry}}/{file}",
            file=[
                "gene.counts.processed.noComBat.tsv",
                "gene.counts.processed.tsv",
                "gene.counts.scaled.normalized.bed.gz",
                "gene.counts.scaled.normalized.bed.gz.tbi",
            ],
        ),
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/ancestry_expr.R",
        data_dir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/",
        level="gene",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        mkdir -p {params.data_dir}{wildcards.ancestry}
        Rscript {params.script} \
            --data_dir {params.data_dir} \
            --ancestry {wildcards.ancestry} \
            --subj_list {input[4]} \
            --level {params.level}
        bgzip {params.data_dir}{wildcards.ancestry}/{params.level}.counts.scaled.normalized.bed
        tabix -p bed {params.data_dir}{wildcards.ancestry}/{params.level}.counts.scaled.normalized.bed.gz
        """


rule cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{num_hcp}hcp_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        script="scripts/cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


# I had to change output file name to make it unambiguous with rule cov
rule ancestry_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        script="scripts/ancestry_cov.R",
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


rule fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule ancestry_fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=16,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/",
        script="scripts/call_nominal.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


rule ancestry_merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP/{file}",
            file=[
                "all_assoc.txt.gz",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/",
        script="scripts/call_nominal.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


# Had trouble with 500 chunks, some chunks do not have any variants
rule fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/90hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=4,
    params:
        outdir=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp"
        ),
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 300 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule ancestry_fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 300 \
                          --seed 123 \
                          --window 1e6 \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_perm:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 301, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 301, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/",
        script="scripts/call_perm.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


rule ancestry_merge_perm:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 301, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 301, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/",
        script="scripts/call_perm.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


############################## QTLtools conditional pass ##############################
rule permutations_all_threshold:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/all.chunks.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/permutations_all_threshold.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        script="scripts/permutations_all_threshold.R",
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --outdir {params.outdir}
        gzip {params.outdir}permutations_all_threshold.txt
        """


rule make_qtltools_bed:
    input:
        fastqtl_expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/qtltools.gene.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/qtltools.gene.counts.scaled.normalized.bed.gz.tbi",
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        zcat {input.fastqtl_expr} | awk '{{ $4=$4" . +"; print $0 }}' | tr " " "\t" | bgzip -c > {output[0]}
        tabix -p bed {output[0]}
        """


rule conditional:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/qtltools.gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/90hcp_cov.txt",
        threshold="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/permutations_all_threshold.txt.gz",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/chunk{chunk}.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/chunk{chunk}.txt.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/",
    shell:
        """
        mkdir -p {params.outdir}
        {config[QTLTOOLS]} cis \
            --vcf {input.geno} \
            --bed {input.expr} \
            --cov {input.cov} \
            --mapping {input.threshold} \
            --chunk {wildcards.chunk} 300 \
            --out {output[0]} \
            --exclude-samples {input.rel_file} \
            --seed 123 \
            --window 1000000 \
            --normal
        touch {output[1]}
        """


rule get_top_variants:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/chunk{chunk}.txt.done",
            chunk=np.arange(1, 301, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/conditional_full.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/conditional_top_variants.txt",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_conditional_90hcp/",
    shell:
        """
        cat {params.outdir}chunk*.txt > {params.outdir}conditional_full.txt
        cat {params.outdir}conditional_full.txt | awk '{{ if ($19 == 1) print $0}}' > {params.outdir}conditional_top_variants.txt
        """


################################### susie finemapping ###################################
rule vcf_to_dosage:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{file}",
            file=[
                "4_columns.tsv",
                "sample_list.tsv",
                "header.tsv",
                "header_row.tsv.gz",
                "dose_matrix.tsv.gz",
                "filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz",
                "filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load htslib/1.9

        # Extract header
        printf 'CHROM\nPOS\nREF\nALT\n' > {output[0]}
        bcftools query -l {input.vcf} > {output[1]}
        cat {output[0]} {output[1]} > {output[2]}
        {config[CSVTK]} transpose {output[2]} -T | gzip > {output[3]}

        # Extract dosage and merge
        bcftools +dosage {input.vcf} -- -t GT | tail -n+2 | gzip > {output[4]}
        zcat {output[3]} {output[4]} | bgzip > {params.prefix}.dose.tsv.gz
        tabix -s1 -b2 -e2 -S1 {params.prefix}.dose.tsv.gz
        """


rule make_susie_meta:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/phenotype_meta.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/sample_meta.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/make_susie_meta.R",
        qtl_group="mixed_ciseqtl_90hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --bed {input[0]} \
            --qtl_group {params.qtl_group} \
            --exclude_rel {input[1]}
        """


rule make_susie_expr:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.susie.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/make_susie_expr.R",
        outname="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.susie.tsv",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --input {input[0]} \
            --outname {params.outname} \
            --exclude_rel {input[1]}
        """


rule run_susie:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.processed.susie.tsv",
        pheno_meta=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/phenotype_meta.tsv"
        ),
        sample_meta=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/sample_meta.tsv"
        ),
        sig_pheno="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_perm_90hcp/sig_pheno.txt",
        genotype_matrix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz",
        covariates=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/90hcp_cov_629.txt"
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm.{{batch}}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
        ),
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=120,
    params:
        script="scripts/run_susie_customized.R",
        cisdistance=1000000,
        permuted="true",
        qtl_group="mixed_ciseqtl_90hcp_perm",
        out_dir=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/"
        ),
        quant_method="gene_counts",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        mkdir -p {params.out_dir}

        Rscript {params.script} \
            --expression_matrix {input.expr} \
            --phenotype_meta {input.pheno_meta} \
            --sample_meta {input.sample_meta} \
            --sig_pheno {input.sig_pheno} \
            --covariates {input.covariates} \
            --genotype_matrix {input.genotype_matrix} \
            --chunk '{wildcards.batch} {config[SUSIE_N_BATCH]}' \
            --cisdistance {params.cisdistance} \
            --out_prefix {params.out_dir}{params.qtl_group}.{wildcards.batch}_{config[SUSIE_N_BATCH]} \
            --permuted {params.permuted} \
            --qtl_group {params.qtl_group} \
            --quant_method {params.quant_method}
        """


rule merge_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm.{batch}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
            batch=np.arange(1, 501, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.txt | bgzip -c > {params.prefix}.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.cred.txt | bgzip -c > {params.prefix}.cred.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.snp.txt | bgzip -c > {params.prefix}.snp.txt.gz
        """


rule sort_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm_susie_merged.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm_purity_filtered.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/mixed_ciseqtl_90hcp_perm",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        gunzip -c {input[0]} > {params.prefix}_susie_merged.txt
        (head -n 1 {params.prefix}_susie_merged.txt && tail -n +2 {params.prefix}_susie_merged.txt | sort -k3 -k4n ) | bgzip > {params.prefix}_purity_filtered.txt.gz
        """


############################# ancestry susie finemapping #############################
rule vcf_to_dosage_ancestry:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{{ancestry}}/{file}",
            file=[
                "4_columns.tsv",
                "sample_list.tsv",
                "header.tsv",
                "header_row.tsv.gz",
                "dose_matrix.tsv.gz",
                "filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz",
                "filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load htslib/1.9

        # Extract header
        printf 'CHROM\nPOS\nREF\nALT\n' > {output[0]}
        bcftools query -l {input.vcf} > {output[1]}
        cat {output[0]} {output[1]} > {output[2]}
        {config[CSVTK]} transpose {output[2]} -T | gzip > {output[3]}

        # Extract dosage and merge
        bcftools +dosage {input.vcf} -- -t GT | tail -n+2 | gzip > {output[4]}
        zcat {output[3]} {output[4]} | bgzip > {params.prefix}.dose.tsv.gz
        tabix -s1 -b2 -e2 -S1 {params.prefix}.dose.tsv.gz
        """


# Note in Rscript num_hcp <- substring(args$qtl_group, 13, 14)
rule make_susie_meta_ancestry:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.scaled.normalized.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/phenotype_meta_{num_hcp}hcp.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/sample_meta_{num_hcp}hcp.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/make_susie_meta.R",
        qtl_group="{ancestry}_ciseqtl_{num_hcp}hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --bed {input[0]} \
            --qtl_group {params.qtl_group} \
            --exclude_rel {input[1]}
        """


rule make_susie_expr_ancestry:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.susie.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        script="scripts/make_susie_expr.R",
        outname="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.susie.tsv",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript {params.script} \
            --input {input[0]} \
            --outname {params.outname} \
            --exclude_rel {input[1]}
        """


rule run_susie_ancestry:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/gene.counts.processed.susie.tsv",
        pheno_meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/phenotype_meta_{num_hcp}hcp.tsv",
        sample_meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/sample_meta_{num_hcp}hcp.tsv",
        sig_pheno="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_perm_{num_hcp}HCP/sig_pheno.txt",
        genotype_matrix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeGeneOutlier.dose.tsv.gz",
        covariates="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{ancestry}/{num_hcp}HCP_cov_final.txt",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{{ancestry}}_ciseqtl_{{num_hcp}}hcp_perm.{{batch}}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
        ),
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=120,
    params:
        script="scripts/run_susie_customized.R",
        cisdistance=1000000,
        permuted="true",
        qtl_group="{ancestry}_ciseqtl_{num_hcp}hcp_perm",
        out_dir=(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/"
        ),
        quant_method="gene_counts",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        mkdir -p {params.out_dir}

        Rscript {params.script} \
            --expression_matrix {input.expr} \
            --phenotype_meta {input.pheno_meta} \
            --sample_meta {input.sample_meta} \
            --sig_pheno {input.sig_pheno} \
            --covariates {input.covariates} \
            --genotype_matrix {input.genotype_matrix} \
            --chunk '{wildcards.batch} {config[SUSIE_N_BATCH]}' \
            --cisdistance {params.cisdistance} \
            --out_prefix {params.out_dir}{params.qtl_group}.{wildcards.batch}_{config[SUSIE_N_BATCH]} \
            --permuted {params.permuted} \
            --qtl_group {params.qtl_group} \
            --quant_method {params.quant_method}
        """


rule merge_susie_ancestry:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{{ancestry}}_ciseqtl_{{num_hcp}}hcp_perm.{batch}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
            batch=np.arange(1, 501, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{{ancestry}}_ciseqtl_{{num_hcp}}hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{ancestry}_ciseqtl_{num_hcp}hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.txt | bgzip -c > {params.prefix}.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.cred.txt | bgzip -c > {params.prefix}.cred.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.snp.txt | bgzip -c > {params.prefix}.snp.txt.gz
        """


rule sort_susie_ancestry:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{{ancestry}}_ciseqtl_{{num_hcp}}hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz", "snp.txt.gz"],
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{ancestry}_ciseqtl_{num_hcp}hcp_perm_susie_merged.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{ancestry}_ciseqtl_{num_hcp}hcp_perm_purity_filtered.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/susie_finemap/{ancestry}_ciseqtl_{num_hcp}hcp_perm",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        gunzip -c {input[0]} > {params.prefix}_susie_merged.txt
        (head -n 1 {params.prefix}_susie_merged.txt && tail -n +2 {params.prefix}_susie_merged.txt | sort -k3 -k4n ) | bgzip > {params.prefix}_purity_filtered.txt.gz
        """

# --ld-vcf has problem with marker ID
# all hg19 chr:pos in METAL file
# can't find matching ones in VCF
rule locuszoom:
    input:
        metal="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/{ancestry}_nominal_{num_hcp}HCP/ENSG00000117899.all_assoc_METAL.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/locuszoom_{ancestry}_{num_hcp}HCP_chr15_81279136.pdf"
    resources:
        mem_gb=4,
        time_min=60,
        num_cores=2
    params:
        out="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/locuszoom_{ancestry}_{num_hcp}HCP"
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        module load python # 2.7.5
        module load htslib/1.12
        module load plink/1.07
        
        # find the FDR closet to 0.05, then signif cutoff is -log10(npval)
        if [[ "{wildcards.ancestry}" == "afr" ]]; then
            signif=4.502014
            pop="AFR"
        elif [[ "{wildcards.ancestry}" == "amr" ]]; then
            signif=3.976155
            pop="AMR"
        elif [[ "{wildcards.ancestry}" == "eur" ]]; then
            signif=3.648817
            pop="EUR"
        fi
        echo ${{pop}}
        echo ${{signif}}
        /u/project/gandalm/shared/apps/locuszoom/bin/locuszoom \
            theme=publication \
            --cache None \
            --no-date \
            --plotonly \
            --gene-table gencode \
            --build hg19 \
            --metal {input.metal} \
            --refsnp chr15:81279136 \
            --flank 50kb \
            --pop ${{pop}} \
            --source 1000G_Nov2014 \
            --prefix {params.out} \
            signifLine=\"${{signif}}\" \
            signifLineColor=\"gray\" \
            signifLineWidth=\"2\" \
            showRecomb=FALSE \
            width=10 \
            height=5 \
            showGenes=FALSE
        """
################################### Ancestry effect size ###################################
rule make_effect_size_scatter_eur_amr:
    input:
        pop1="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/eur_nominal_50HCP/all_assoc.txt.gz",
        pop2="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/amr_nominal_15HCP/all_assoc.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_eur_amr.png",
    resources:
        mem_gb=6,
        time_min=120,
        num_cores=4,
    params:
        script="scripts/nominal_effect_size_scatter_plot.R",
        pop12_group="EUR, AMR",
        pop1_group="EUR only",
        pop2_group="AMR only",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --pop1 {input.pop1} \
            --pop2 {input.pop2} \
            --out {output[0]} \
            --pop12_group "{params.pop12_group}" \
            --pop1_group "{params.pop1_group}" \
            --pop2_group "{params.pop2_group}"
        """


rule make_effect_size_scatter_eur_afr:
    input:
        pop1="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/eur_nominal_50HCP/all_assoc.txt.gz",
        pop2="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/afr_nominal_25HCP/all_assoc.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_eur_afr.png",
    resources:
        mem_gb=6,
        time_min=120,
        num_cores=4,
    params:
        script="scripts/nominal_effect_size_scatter_plot.R",
        pop12_group="EUR, AFR",
        pop1_group="EUR only",
        pop2_group="AFR only",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --pop1 {input.pop1} \
            --pop2 {input.pop2} \
            --out {output[0]} \
            --pop12_group "{params.pop12_group}" \
            --pop1_group "{params.pop1_group}" \
            --pop2_group "{params.pop2_group}"
        """


rule make_effect_size_scatter_afr_amr:
    input:
        pop1="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/afr_nominal_25HCP/all_assoc.txt.gz",
        pop2="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/amr_nominal_15HCP/all_assoc.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/scatter_afr_amr.png",
    resources:
        mem_gb=6,
        time_min=120,
        num_cores=4,
    params:
        script="scripts/nominal_effect_size_scatter_plot.R",
        pop12_group="AFR, AMR",
        pop1_group="AFR only",
        pop2_group="AMR only",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --pop1 {input.pop1} \
            --pop2 {input.pop2} \
            --out {output[0]} \
            --pop12_group "{params.pop12_group}" \
            --pop1_group "{params.pop1_group}" \
            --pop2_group "{params.pop2_group}"
        """


rule scatter_eur_afr:
    input:
        pop1="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/eur_perm_50HCP/sig_pheno.txt",
        pop2="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/afr_nominal_25HCP/all_assoc.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/perm_scatter_eur_afr.png",
    resources:
        mem_gb=6,
        time_min=120,
        num_cores=4,
    params:
        script="scripts/scatter_plot.R",
        pop1_group="EUR",
        pop2_group="AFR",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --pop1 {input.pop1} \
            --pop2 {input.pop2} \
            --out {output[0]} \
            --pop1_group "{params.pop1_group}" \
            --pop2_group "{params.pop2_group}"
        """


rule scatter_eur_amr:
    input:
        pop1="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/eur_perm_50HCP/sig_pheno.txt",
        pop2="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/amr_nominal_15HCP/all_assoc.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/figures/perm_scatter_eur_amr.png",
    resources:
        mem_gb=6,
        time_min=120,
        num_cores=4,
    params:
        script="scripts/scatter_plot.R",
        pop1_group="EUR",
        pop2_group="AMR",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript {params.script} \
            --pop1 {input.pop1} \
            --pop2 {input.pop2} \
            --out {output[0]} \
            --pop1_group "{params.pop1_group}" \
            --pop2_group "{params.pop2_group}"
        """


################################### Torus functional enrichment ###################################
rule make_annot_ensembl:
    input:
        ref="/u/project/gandalm/shared/refGenomes/ensembl_regulatory_build/homo_sapiens.GRCh37.Regulatory_Build.regulatory_features.20201218.gff.gz",
        coord="/u/project/gandalm/cindywen/isoform_twas/sLDSC/data/mixed_variant_coord.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_chr{chr}.tsv",
    resources:
        mem_gb=4,
        time_min=240,
        num_cores=6,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript scripts/make_annot.R \
            {input.ref} \
            {input.coord} \
            {wildcards.chr} \
            {params.prefix}
        """


rule run_vep:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.vcf.gz",
        gtf_bgz="/u/project/gandalm/shared/refGenomes/hg19/Gencode/v33/gencode.v33lift37.annotation.sorted.gtf.gz",
        gtf_tbi="/u/project/gandalm/shared/refGenomes/hg19/Gencode/v33/gencode.v33lift37.annotation.sorted.gtf.gz.tbi",
        fasta_bgz="/u/project/gandalm/shared/refGenomes/hg19/Gencode/GRCh37.primary_assembly.genome.fa.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_effect_output.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_effect_output.txt.gz_summary.html",
    resources:
        mem_gb=4,
        time_min=240,
        num_cores=6,
    shell:
        """
        source /u/local/apps/anaconda3/2019.03/bin/activate ensembl-vep
        vep -i {input.vcf} \
            --cache \
            --dir_cache /u/project/gandalm/shared/refGenomes/ensembl_vep_cache/ \
            --gtf {input.gtf_bgz} \
            --fasta {input.fasta_bgz} \
            --output_file {output[0]} \
            --fork {resources.num_cores} \
            --tab \
            --format vcf \
            --assembly GRCh37 \
            --compress_output gzip \
            --offline
        """


rule select_consequences:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_effect_output.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_effect_output_info.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        zcat {input[0]} | grep -v '^##' | awk '$18 == "gencode.v33lift37.annotation.sorted.gtf.gz" {{print $1"\t"$7}}' > {output[0]}
        """


rule add_vep:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_effect_output_info.txt",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_chr{chr}.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep_chr{chr}.tsv",
    resources:
        mem_gb=4,
        time_min=720,
        num_cores=6,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript scripts/add_vep.R \
            --vep {input[0]} \
            --ensembl {input[1]} \
            --out {output[0]}
        """


rule merge_annot_ensembl_vep:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep_chr{chr}.tsv",
            chr=np.arange(1, 23, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep.txt.gz",
    resources:
        mem_gb=4,
        time_min=240,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep",
    shell:
        """
        printf 'SNP\tTF_binding_site_d\tpromoter_flanking_region_d\tpromoter_d\topen_chromatin_region_d\tenhancer_d\tCTCF_binding_site_d\tsynonymous_variant_d\tstop_gained_d\tsplice_region_variant_d\tsplice_donor_variant_d\tsplice_acceptor_variant_d\tnon_coding_transcript_exon_variant_d\tmissense_variant_d\tintron_variant_d\tframeshift_variant_d\t5_prime_UTR_variant_d\t3_prime_UTR_variant_d\n' > {params.prefix}.txt
        for i in {{1..22}};do
        cat {params.prefix}_chr${{i}}.tsv >> {params.prefix}.txt
        done
        gzip {params.prefix}.txt
        """


# # Here we calculate se_beta for all nominal associations
# rule fastqtl_calculate_sebeta:
#     input:
#         "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/chunk{chunk}.txt.gz",
#         # "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/significant_assoc.txt",
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_chunk{chunk}.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_chunk{chunk}.txt.gz.done",
#     resources:
#         mem_gb=6,
#         time_min=480,
#         num_cores=12,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_chunk{chunk}",
#         # num_comparisons=187820592 # number of lines in all.chunks.txt.gz
#     shell:
#         """
#         . /u/local/Modules/default/init/modules.sh
#         module load R/3.6.0
#         Rscript scripts/calculate_sebeta.R \
#             --input {input[0]} \
#             --output {params.prefix}.txt
#         gzip {params.prefix}.txt
#         touch {params.prefix}.txt.gz.done
#         """


# rule merge_sebeta:
#     input:
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_chunk{chunk}.txt.gz",
#             chunk=np.arange(1, 101, 1),
#         ),
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_chunk{chunk}.txt.gz.done",
#             chunk=np.arange(1, 101, 1),
#         ),
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_all_chunks.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_all_chunks.txt.gz.done",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/sebeta_",
#     shell:
#         """
#         zcat {params.prefix}chunk*.txt.gz | gzip -c > {params.prefix}all_chunks.txt.gz
#         touch {params.prefix}all_chunks.txt.gz.done
#         """


# The following three rules are basically running run_FastQTL_threaded.py in GTEx's version of FastQTL
rule gtex_fastqtl:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/gene.counts.scaled.normalized.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeGeneOutlier.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.log",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex_chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp",
    shell:
        """
        # mkdir -p {params.outdir}
        . /u/local/Modules/default/init/modules.sh
        module load gcc/9.3.0
        /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/bin/fastQTL.static \
            --vcf {input.geno} \
            --bed {input.expr} \
            --window 1e6 \
            --cov {input.cov} \
            --seed 123 \
            --exclude-samples {input.rel_file} \
            --chunk {wildcards.chunk} 100 \
            --out {output[0]} \
            --log {output[1]} \
            --normal
        touch {output[0]}.done
        """


rule write_chunk_log_list:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{{num_hcp}}hcp/gtex_chunk{chunk}.{file}",
            file=["txt.gz", "log", "txt.gz.done"],
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/list_log.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        for i in {{1..100..1}}; do
        echo /u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp/gtex_chunk${{i}}.txt.gz >> {output[0]}
        echo /u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp/gtex_chunk${{i}}.log >> {output[1]}
        done
        """


rule gtex_merge_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/list_log.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/gtex.txt.gz",
    resources:
        num_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_{num_hcp}hcp/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        module load R/3.6.0
        python3 /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/python/merge_chunks.py \
            {input[0]} \
            {input[1]} \
            gtex \
            --fdr 0.05 \
            -o {params.outdir} \
        """


rule make_torus_input:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/gtex.allpairs.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/gtex.allpairs.torus.txt.gz",
    resources:
        mem_gb=4,
        time_min=120,
    shell:
        """
        zcat {input[0]} | tail -n+2 | awk '{{print $1,$2,$3,$7,$8,$9}}' OFS=' ' | gzip > {output[0]}
        """


rule run_torus:
    input:
        annot="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep.txt.gz",
        fastqtl="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/gtex.allpairs.torus.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/eqtl_new/results/mixed_nominal_90hcp/torus.out",
    resources:
        mem_gb=4,
        time_min=720,
        num_cores=10,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        /u/project/gandalm/shared/apps/torus-master/src/torus \
            -d {input.fastqtl} \
            --fastqtl \
            -est \
            -annot {input.annot} \
            > {output[0]}
        """


################################### PAINTOR multi-ethnic fine-mapping ###################################
# Variant coord for ALL is in sLDSC dir
rule make_eur_coord:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/filtered.hg19.sorted.vcf.gz",
    output:
        info="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_info.tsv",
        coord="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_coord.tsv",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        script="/u/project/gandalm/cindywen/isoform_twas/sLDSC/code/scripts/variant_coord.R",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load R/3.6.0
        bcftools query -f '%CHROM %POS %ID\n' {input.vcf} -o {output.info}
        Rscript {params.script} \
            --info {output.info} \
            --out {output.coord}
        """


rule make_eur_annot:
    input:
        ref="/u/project/gandalm/shared/refGenomes/ensembl_regulatory_build/homo_sapiens.GRCh37.Regulatory_Build.regulatory_features.20201218.gff.gz",
        coord="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_coord.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot_chr{chr}.tsv",
    resources:
        mem_gb=4,
        time_min=240,
        num_cores=6,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript scripts/make_annot.R \
            {input.ref} \
            {input.coord} \
            {wildcards.chr} \
            {params.prefix}
        """


rule merge_eur_annot:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot_chr{chr}.tsv",
            chr=np.arange(1, 23, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot.txt.gz",
    resources:
        mem_gb=4,
        time_min=240,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/eur_variant_annot",
    shell:
        """
        printf 'SNP\tTF_binding_site_d\tpromoter_flanking_region_d\tpromoter_d\topen_chromatin_region_d\tenhancer_d\tCTCF_binding_site_d\n' > {params.prefix}.txt
        for i in {{1..22}};do
        cat {params.prefix}_chr${{i}}.tsv >> {params.prefix}.txt
        done
        gzip {params.prefix}.txt
        """
