from os.path import join
import os
import numpy as np
import sys


configfile: "config.yaml"


# WINDOWS = ["1e5", "1e6"]
WINDOWS = ["1e6"]

"""
rules:
STAR
    - index: generate STAR index with Gencode v33, with annotation GTF as recommended
    - See step 1-5 for running STAR, 1st and 2nd pass, WASP filter; Leafcutter bam2junc
Leafcutter
    - cluster: note some output files are not specified here
    - remove_chr
    - write_chr_blacklist
    - pheno_prep: note some output files are not specified here
    - bgzip_tabix
    - concat
    - pheno_process
    - cov
ALL
    - fastqtl_nominal
    - merge_nominal
    - call_nominal
    - fastqtl_perm
    - merge_perm
    - call_perm
Separate ancestry
    - ancestry_pheno
    - ancestry_cov
    - ancestry_fastqtl_nominal
    - ancestry_merge_nominal
    - ancestry_call_nominal
    - ancestry_fastqtl_perm
    - ancestry_merge_perm
    - ancestry_call_perm
Intron annotation
    - prepare_leafviz_annot
    - annotate_intron
    - summarise_annot
GTEx way
    - prepare_leafviz_annot_exons: add gene_id in all_exons file
    - map_clusters_to_genes
    - make_grp_and_bed_file
    - fastqtl_grp_perm
    - gtex_grp_perm_write_chunk_log_list
    - gtex_grp_perm_merge_chunk
Conditional 
    - permutations_all_threshold: compute a npval threshold for all tested introns
    - conditional: run QTLtools conditional pass
    - get_top_variants: get top variants per rank
susie finemapping
    - vcf_to_dosage
    - make_susie_meta
    - make_susie_expr
    - run_susie
    - merge_susie
    - sort_susie
Torus: functional enrichment
    - (fastqtl_calculate_sebeta)
    - (merge_beta)
    - gtex_fastqtl
    - gtex_write_chunk_log_list
    - gtex_merge_chunk
    - make_torus_input
    - run_torus
qvalue pi0 prep
    - qvalue_pi0_prep_chunk
    - qvalue_pi0_prep_merge
trimeser-specific
    - tri_cov
    - (celine ran sQTL mapping)
sex-specific
    - sex_cov
    - (Pinto Lab)
"""


rule all:
    input:
        # "/u/project/gandalm/cindywen/isoform_twas/star/gencodev33_STARindex/SAindex",
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/{file}",
        #     num_hcp=np.arange(5, 51, 5),
        #     window=WINDOWS,
        #     file=[
        #         "all_assoc.txt",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/all_assoc.txt.gz",
        #     zip,
        #     best_num_hcp=[40],
        #     window=WINDOWS,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/sig_pheno.txt",
        #     zip,
        #     best_num_hcp=[40],
        #     window=WINDOWS,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/{file}",
        #     ancestry="eur",
        #     num_hcp=np.arange(5, 51, 5),
        #     window=WINDOWS,
        #     file=[
        #         "all_assoc.txt",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/{file}",
        #     ancestry=["amr", "afr"],
        #     num_hcp=np.arange(2, 21, 2),
        #     window=WINDOWS,
        #     file=[
        #         "all_assoc.txt",
        #         "significant_assoc.txt",
        #         "significant_feature_count.txt",
        #     ],
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry="eur",
        #     num_hcp=20,
        #     window=WINDOWS,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry="amr",
        #     num_hcp=10,
        #     window=WINDOWS,
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/{file}",
        #     file=["all_assoc.txt.gz", "sig_pheno.txt"],
        #     ancestry="afr",
        #     num_hcp=12,
        #     window=WINDOWS,
        # ),
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/all.introns.RData",
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/group.perm.txt.gz",
        #     num_hcp=40,
        #     window="1e6",
        # ),
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/group.perm.genes.txt.gz",
        #     num_hcp=40,
        #     window="1e6",
        # ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/{file}",
            file=["conditional_full.txt", "conditional_top_variants.txt"],
        ),
        # "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm_purity_filtered.txt.gz",
        # "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/torus.out",
        # "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/qvalue.done",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri_specific/tri{tri}_{num_hcp}HCP_cov.txt",
            tri=[1, 2],
            num_hcp=np.arange(5, 51, 5),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/sex_specific/{sex}_{num_hcp}HCP_cov.txt",
            sex=["m", "f"],
            num_hcp=np.arange(5, 51, 5),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{tri}_cissqtl_{num_hcp}hcp_perm_purity_filtered.txt.gz",
            zip,
            tri=[1, 2],
            num_hcp=[15, 10],
        ),


rule index:
    input:
        config["GENOME_FASTA"],
        config["GTF_FILE"],
    output:
        "/u/project/gandalm/cindywen/isoform_twas/star/gencodev33_STARindex/SAindex",
    resources:
        mem_gb=4,
        time_min=120,
        num_cores=8,
    shell:
        """
        {config[STAR]} --runThreadN {resources.num_cores} \
            --runMode genomeGenerate \
            --genomeDir /u/project/gandalm/cindywen/isoform_twas/star/gencodev33_STARindex \
            --genomeFastaFiles {config[GENOME_FASTA]} \
            --sjdbGTFfile {config[GTF_FILE]} \
            --sjdbOverhang 100
        """


rule cluster:
    input:
        "juncfiles.txt",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_{file}",
            file=[
                "pooled",
                "refined",
                "sortedlibs",
                "perind.counts.gz",
                "perind_numers.counts.gz",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=150,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/2.7.16
        mkdir -p {params.outdir}
        python config[LEAFCUTTER]clustering/leafcutter_cluster.py \
            -j {input[0]} \
            -m 50 \
            -l 500000 \
            -r {params.outdir}
        """


rule remove_chr:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz",
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        zcat {input[0]} | sed '2, $s/chr//' | gzip -c > {output[0]}
        """


rule write_chr_blacklist:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz",
    output:
        "chr_remove.txt",
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript scripts/write_chr_blacklist.R \
            --input {input[0]} \
            --output {output[0]}
        """


rule pheno_prep:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz",
        "chr_remove.txt",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz{file}",
            file=["_prepare.sh", ".PCs", ".ave"],
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.phen_chr{chr_num}",
            chr_num=np.arange(1, 23, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}",
            chr_num=np.arange(1, 23, 1),
        ),
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/2.7.16
        module load htslib/1.9
        python config[LEAFCUTTER]scripts/prepare_phenotype_table.py \
            {input[0]} \
            --ChromosomeBlackList {input[1]}
        """


rule bgzip_tabix:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz_prepare.sh",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}",
            chr_num=np.arange(1, 23, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}.gz",
            chr_num=np.arange(1, 23, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}.gz.tbi",
            chr_num=np.arange(1, 23, 1),
        ),
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        bash {input[0]}
        """


rule concat:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}.gz",
            chr_num=np.arange(1, 23, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_chr{chr_num}.gz.tbi",
            chr_num=np.arange(1, 23, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all.gz.tbi",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        for i in {{1..22}}; do
        zcat {params.prefix}_chr${{i}}.gz | bgzip -f >> {params.prefix}_all.gz
        done
        tabix -p bed {params.prefix}_all.gz
        """


rule pheno_process:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all.gz.tbi",
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/geno_subj_{study}.txt",
            study=["walker", "obrien", "werling", "hdbr", "libd"],
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_{file}",
            file=[
                "fixSubj.tsv",
                "fixSubj_combat.tsv",
                "fixSubj_combat.bed.gz",
                "fixSubj_combat.bed.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all",
        geno_subj_dir="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/geno_subj/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9
        Rscript scripts/pheno_process.R \
            --prefix {params.prefix} \
            --geno_subj_dir {params.geno_subj_dir}
        bgzip -f {params.prefix}_fixSubj_combat.bed
        tabix -p bed {params.prefix}_fixSubj_combat.bed.gz
        """


rule cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{num_hcp}hcp_cov.txt",
    resources:
        mem_gb=8,
        time_min=120,
    params:
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        mkdir -p {params.outdir}
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/cov.R \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


############################## ALL ##############################
rule fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=150,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window {wildcards.window} \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{{num_hcp}}hcp_{{window}}/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{{num_hcp}}hcp_{{window}}/{file}",
            file=[
                "all_assoc.txt",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=12,
        time_min=200,
        num_cores=16,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_nominal.R \
            --outdir {params.outdir}
        # gzip {params.outdir}all_assoc.txt
        # bgzip {params.outdir}all_assoc.txt
        """


rule fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{best_num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=1440,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window {wildcards.window} \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule merge_perm:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{{best_num_hcp}}hcp_{{window}}/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{{best_num_hcp}}hcp_{{window}}/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{{best_num_hcp}}hcp_{{window}}/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_{best_num_hcp}hcp_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_perm.R \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


############################## Ancestry ##############################
rule ancestry_pheno:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_{file}",
            file=[
                "fixSubj.tsv",
                "fixSubj_combat.tsv",
                "fixSubj_combat.bed.gz",
                "fixSubj_combat.bed.gz.tbi",
            ],
        ),
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/ancestry_list/list.{ancestry}.tsv",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{{ancestry}}/{file}",
            file=[
                "lc.tsv",
                "lc_combat.tsv",
                "lc_combat.bed.gz",
                "lc_combat.bed.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/",
        combined_dir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9

        mkdir -p {params.outdir}
        Rscript scripts/ancestry_lc.R \
            --combined_dir {params.combined_dir} \
            --subj_list {input[4]} \
            --outdir {params.outdir}
        bgzip {params.outdir}lc_combat.bed
        tabix -p bed {params.outdir}lc_combat.bed.gz
        """


rule ancestry_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/lc_combat.tsv",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/ancestry_cov.R \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir}
        """


rule ancestry_fastqtl_nominal:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/lc_combat.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=180,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 100 \
                          --seed 123 \
                          --window {wildcards.window} \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


rule ancestry_merge_nominal:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP_{{window}}/chunk{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP_{{window}}/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/all.chunks.txt.gz",
    resources:
        mem_gb=24,
        time_min=160,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_nominal:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_nominal_{{num_hcp}}HCP_{{window}}/{file}",
            file=[
                "all_assoc.txt",
                "significant_assoc.txt",
                "significant_feature_count.txt",
            ],
        ),
    resources:
        mem_gb=12,
        time_min=200,
        num_cores=16,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_nominal_{num_hcp}HCP_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_nominal.R \
            --outdir {params.outdir}
        # gzip {params.outdir}all_assoc.txt
        """


# ran in bash
# some of the 1000 chunks do not have genotype
# snakemake will terminate, bash won't
rule ancestry_fastqtl_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/lc_combat.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{ancestry}/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{ancestry}/{num_hcp}HCP_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        # "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}",
    shell:
        """
        mkdir -p {params.outdir}
        {config[FASTQTL]} --vcf {input.geno} \
                          --bed {input.expr} \
                          --cov {input.cov} \
                          --permute 1000 10000 \
                          --out {output[0]} \
                          --chunk {wildcards.chunk} 1000 \
                          --seed 123 \
                          --window {wildcards.window} \
                          --normal \
                          --exclude-samples {input.rel_file}
        touch {output[0]}.done
        """


# first check 2000 files exist, then
# manually changed empty chunk files to .txt
# skip those, or would get error
# also have to comment out the .txt.gz input
# delete the .txt.gz files
rule ancestry_merge_perm:
    input:
        # expand(
        #     "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP_{{window}}/chunk{chunk}.txt.gz",
        #     chunk=np.arange(1, 1001, 1),
        # ),
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP_{{window}}/chunk{chunk}.txt.gz.done",
            chunk=np.arange(1, 1001, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/all.chunks.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/",
    shell:
        """
        zcat {params.outdir}chunk*.txt.gz | gzip -c > {params.outdir}all.chunks.txt.gz
        """


rule ancestry_call_perm:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/all.chunks.txt.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{{ancestry}}_perm_{{num_hcp}}HCP_{{window}}/{file}",
            file=["all_assoc.txt.gz", "sig_pheno.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/{ancestry}_perm_{num_hcp}HCP_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/call_perm.R \
            --outdir {params.outdir}
        gzip {params.outdir}all_assoc.txt
        """


############################## Intron annotation ##############################
rule prepare_leafviz_annot:
    input:
        config["GTF_FILE"],
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_{file}",
            file=[
                "all_introns.bed.gz",
                "threeprime.bed.gz",
                "fiveprime.bed.gz",
                "all_exons.txt.gz",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/",
        out_prefix="gencode_v33lift37",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load perl/5.10.1
        mkdir -p {params.out_dir}
        config[LEAFCUTTER]leafviz/gtf2leafcutter.pl \
            -o {params.out_dir}{params.out_prefix} \
            {config[GTF_FILE]}
        """


rule annotate_intron:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_{file}",
            file=[
                "all_introns.bed.gz",
                "threeprime.bed.gz",
                "fiveprime.bed.gz",
                "all_exons.txt.gz",
            ],
        ),
        counts="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind_numers.counts.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafviz.RData",
    resources:
        mem_gb=4,
        num_cores=10,
        time_min=480,
    params:
        annot="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37",
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript scripts/annotate_intron.R \
            --annot {params.annot} \
            --counts {input.counts} \
            --outdir {params.outdir}
        """


rule summarise_annot:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafviz.RData",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/all.introns.RData",
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=360,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript scripts/summarise_annot.R \
            --annot {input[0]}
        """


############################## GTEx way, grouped permutations ##############################
# original all_exons file has only gene names which have duplicates, here add gene ID
# this is very slow. Exon big file
# rule make_all_exons:
#     input:
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_all_exons.txt.gz",
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_all_exons_withENSG.txt.gz",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_all_exons_withENSG"
#     shell:
#         """
#         . /u/local/Modules/default/init/modules.sh
#         module load R/3.6.0
#         module load htslib/1.9
#         Rscript scripts/make_all_exons.R \
#             --out {params.prefix} \
#             --gtf {config[GTF_FILE]}
#         bgzip {params.prefix}.txt
#         """


# Use new leafcutter script
rule prepare_leafviz_annot_exons:
    input:
        config["GTF_FILE"],
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_withENSG_all_exons.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/",
        out_prefix="gencode_v33lift37_withENSG",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load perl/5.10.1
        mkdir -p {params.out_dir}
        scripts/gtf2leafcutter_all_exons_ensg.pl \
            -o {params.out_dir}{params.out_prefix} \
            {config[GTF_FILE]}
        """


rule map_clusters_to_genes:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/gencode_v33lift37_withENSG_all_exons.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafcutter_clusters_to_genes.txt",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot",
        out_file="leafcutter_clusters_to_genes.txt",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        Rscript scripts/map_clusters_to_genes.R \
            --intron_counts_file {input[1]} \
            --exon_file {input[0]} \
            --output_dir {params.out_dir} \
            --output_name {params.out_file}
        """


# grouping file: intron to gene mappings
# BED file: add gene to intron ID, use gene coordinate as feature coordinate
rule make_grp_and_bed_file:
    input:
        cluster2gene="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafcutter_clusters_to_genes.txt",
        bed="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        info="/u/project/gandalm/cindywen/isoform_twas/salmon/gencode.v33lift37.annotation.gene.info.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafcutter.phenotype_groups.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_grp.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_grp.bed.gz.tbi",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        bed_out="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_grp.bed",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/3.6.0
        module load htslib/1.9
        Rscript scripts/make_grp_and_bed_file.R \
            --cluster_to_gene {input.cluster2gene} \
            --bed {input.bed} \
            --info {input.info} \
            --grp {output[0]} \
            --bed_out {params.bed_out}
        bgzip {params.bed_out}
        tabix -p bed {params.bed_out}.gz
        """


# Celine ran this in bash script
# won't terminate if no genotype in chunk
rule run_fastqtl_grp_perm:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_grp.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
        grp_file="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/leafcutter.phenotype_groups.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/chunk{chunk}.log",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=720,
        num_cores=10,
    params:
        out_dir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/",
    shell:
        """
        mkdir -p {params.out_dir}
        . /u/local/Modules/default/init/modules.sh
        module load gcc/9.3.0
        mkdir -p {params.out_dir}
        /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/bin/fastQTL.static \
            --vcf {input.geno} \
            --bed {input.expr} \
            --window {wildcards.window} \
            --cov {input.cov} \
            --grp {input.grp_file} \
            --permute 1000 10000 \
            --seed 123 \
            --exclude-samples {input.rel_file} \
            --chunk {wildcards.chunk} 1000 \
            --out {output[0]} \
            --log {output[1]} \
            --normal
        touch {output[2]}
        """


rule gtex_grp_perm_write_chunk_log_list:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{{num_hcp}}hcp_{{window}}/chunk{chunk}.{file}",
            file=["txt.gz", "log", "txt.gz.done"],
            chunk=np.arange(1, 1001, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/list_log.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        for i in {{1..1000..1}}; do
        echo /u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{wildcards.num_hcp}hcp_{wildcards.window}/chunk${{i}}.txt.gz >> {output[0]}
        echo /u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{wildcards.num_hcp}hcp_{wildcards.window}/chunk${{i}}.log >> {output[1]}
        done
        """


rule gtex_grp_perm_merge_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/list_log.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/group.perm.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/group.perm.genes.txt.gz",
    resources:
        num_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_grp_perm_{num_hcp}hcp_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        module load R/3.6.0
        python3 /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/python/merge_chunks.py \
            {input[0]} \
            {input[1]} \
            group.perm \
            --fdr 0.05 \
            -o {params.outdir} \
            --permute
        """


############################## QTLtools conditional pass ##############################
rule permutations_all_threshold:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_40hcp_1e6/all.chunks.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_40hcp_1e6/permutations_all_threshold.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_40hcp_1e6/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/permutations_all_threshold.R \
            --outdir {params.outdir}
        gzip {params.outdir}permutations_all_threshold.txt
        """


rule make_qtltools_bed:
    input:
        fastqtl_expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/qtltools.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/qtltools.bed.gz.tbi",
    resources:
        mem_gb=6,
        time_min=120,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.9
        zcat {input.fastqtl_expr} | awk '{{ $4=$4" . +"; print $0 }}' | tr " " "\t" | bgzip -c > {output[0]}
        tabix -p bed {output[0]}
        """


rule conditional:
    input:
        expr=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/qtltools.bed.gz"
        ),
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/40hcp_cov.txt",
        threshold="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_40hcp_1e6/permutations_all_threshold.txt.gz",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/chunk{chunk}.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/chunk{chunk}.txt.done",
    resources:
        mem_gb=6,
        time_min=480,
        num_cores=6,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/",
    shell:
        """
        mkdir -p {params.outdir}
        {config[QTLTOOLS]} cis \
            --vcf {input.geno} \
            --bed {input.expr} \
            --cov {input.cov} \
            --mapping {input.threshold} \
            --chunk {wildcards.chunk} 100 \
            --out {output[0]} \
            --exclude-samples {input.rel_file} \
            --seed 123 \
            --window 1000000 \
            --normal
        touch {output[1]}
        """


rule get_top_variants:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/chunk{chunk}.txt.done",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/{file}",
            file=["conditional_full.txt", "conditional_top_variants.txt"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_conditional_40hcp_1e6/",
    shell:
        """
        cat {params.outdir}chunk*.txt > {params.outdir}conditional_full.txt
        cat {params.outdir}conditional_full.txt | awk '{{ if ($19 == 1) print $0}}' > {params.outdir}conditional_top_variants.txt
        """


################################### susie finemapping ###################################
rule vcf_to_dosage:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/{file}",
            file=[
                "4_columns_intron.tsv",
                "sample_list_intron.tsv",
                "header_intron.tsv",
                "header_row_intron.tsv.gz",
                "dose_matrix_intron.tsv.gz",
                "filtered.hg19.sorted.removeRel.dose.tsv.gz",
                "filtered.hg19.sorted.removeRel.dose.tsv.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.9
        module load htslib/1.9

        # Extract header
        printf 'CHROM\nPOS\nREF\nALT\n' > {output[0]}
        bcftools query -l {input.vcf} > {output[1]}
        cat {output[0]} {output[1]} > {output[2]}
        {config[CSVTK]} transpose {output[2]} -T | gzip > {output[3]}

        # Extract dosage and merge
        bcftools +dosage {input.vcf} -- -t GT | tail -n+2 | gzip > {output[4]}
        zcat {output[3]} {output[4]} | bgzip > {params.prefix}.dose.tsv.gz
        tabix -s1 -b2 -e2 -S1 {params.prefix}.dose.tsv.gz
        """


rule make_susie_meta:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/phenotype_meta.tsv",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/sample_meta.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_meta.R \
            --bed {input[0]} \
            --exclude_rel {input[1]} \
            --qtl_group "mixed_cissqtl_40hcp_perm"
        """


rule make_susie_expr:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/susie.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_expr.R \
            --input {input[0]} \
            --outname {output[0]} \
            --exclude_rel {input[1]}
        """


rule run_susie:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/susie.tsv",
        pheno_meta="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/phenotype_meta.tsv",
        sample_meta=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/sample_meta.tsv"
        ),
        sig_pheno="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_perm_40hcp_1e6/sig_pheno.txt",
        genotype_matrix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.dose.tsv.gz",
        covariates=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/40hcp_cov.txt"
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm.{{batch}}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
        ),
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=120,
    params:
        cisdistance=1000000,
        permuted="true",
        qtl_group="mixed_cissqtl_40hcp_perm",
        out_dir=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/"
        ),
        quant_method="leafcutter",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        mkdir -p {params.out_dir}

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/run_susie_customized.R \
            --expression_matrix {input.expr} \
            --phenotype_meta {input.pheno_meta} \
            --sample_meta {input.sample_meta} \
            --sig_pheno {input.sig_pheno} \
            --covariates {input.covariates} \
            --genotype_matrix {input.genotype_matrix} \
            --chunk '{wildcards.batch} {config[SUSIE_N_BATCH]}' \
            --cisdistance {params.cisdistance} \
            --out_prefix {params.out_dir}{params.qtl_group}.{wildcards.batch}_{config[SUSIE_N_BATCH]} \
            --permuted {params.permuted} \
            --qtl_group {params.qtl_group} \
            --quant_method {params.quant_method}
        """


rule merge_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm.{batch}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
            batch=np.arange(1, 501, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.12
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.txt | bgzip -c > {params.prefix}.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.cred.txt | bgzip -c > {params.prefix}.cred.txt.gz
        # awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.snp.txt | bgzip -c > {params.prefix}.snp.txt.gz
        """


rule sort_susie:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz"],
        ),
    output:
        # "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm_susie_merged.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm_purity_filtered.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/mixed_cissqtl_40hcp_perm",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.12
        gunzip -c {input[0]} > {params.prefix}_susie_merged.txt
        (head -n 1 {params.prefix}_susie_merged.txt && tail -n +2 {params.prefix}_susie_merged.txt | sort -k3 -k4n ) | bgzip > {params.prefix}_purity_filtered.txt.gz
        """


############################### Torus functional enrichment ###################################
# rule fastqtl_calculate_sebeta:
#     input:
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/chunk{chunk}.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/significant_assoc.txt",
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_chunk{chunk}.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_chunk{chunk}.txt.gz.done",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_chunk{chunk}",
#     shell:
#         """
#         . /u/local/Modules/default/init/modules.sh
#         module load R/3.6.0
#         Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/calculate_sebeta.R \
#             --input {input[0]} \
#             --output {params.prefix}.txt \
#             --sig {input[1]}
#         gzip {params.prefix}.txt
#         touch {params.prefix}.txt.gz.done
#         """


# rule merge_sebeta:
#     input:
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_chunk{chunk}.txt.gz",
#             chunk=np.arange(1, 101, 1),
#         ),
#         expand(
#             "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_chunk{chunk}.txt.gz.done",
#             chunk=np.arange(1, 101, 1),
#         ),
#     output:
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_all_chunks.txt.gz",
#         "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_all_chunks.txt.gz.done",
#     resources:
#         mem_gb=4,
#         time_min=240,
#         num_cores=6,
#     params:
#         prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/sebeta_significant_",
#     shell:
#         """
#         zcat {params.prefix}chunk*.txt.gz | gzip -c > {params.prefix}all_chunks.txt.gz
#         touch {params.prefix}all_chunks.txt.gz.done
#         """


rule gtex_fastqtl:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/leafcutter_perind.counts.nochr.gz.qqnorm_all_fixSubj_combat.bed.gz",
        geno="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/filtered.hg19.sorted.removeRel.vcf.gz",
        cov="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/{num_hcp}hcp_cov.txt",
        rel_file="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/gtex_chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/gtex_chunk{chunk}.log",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/gtex_chunk{chunk}.txt.gz.done",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}",
    shell:
        """
        # mkdir -p {params.outdir}
        . /u/local/Modules/default/init/modules.sh
        module load gcc/9.3.0
        /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/bin/fastQTL.static \
            --vcf {input.geno} \
            --bed {input.expr} \
            --window {wildcards.window} \
            --cov {input.cov} \
            --seed 123 \
            --exclude-samples {input.rel_file} \
            --chunk {wildcards.chunk} 100 \
            --out {output[0]} \
            --log {output[1]} \
            --normal
        touch {output[0]}.done
        """


rule gtex_write_chunk_log_list:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{{num_hcp}}hcp_{{window}}/gtex_chunk{chunk}.{file}",
            file=["txt.gz", "log", "txt.gz.done"],
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/list_log.txt",
    resources:
        mem_gb=4,
        time_min=30,
    shell:
        """
        for i in {{1..100..1}}; do
        echo /u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp_{wildcards.window}/gtex_chunk${{i}}.txt.gz >> {output[0]}
        echo /u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{wildcards.num_hcp}hcp_{wildcards.window}/gtex_chunk${{i}}.log >> {output[1]}
        done
        """


rule gtex_merge_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/list_chunks.txt",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/list_log.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/gtex.allpairs.txt.gz",
    resources:
        num_gb=6,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_{num_hcp}hcp_{window}/",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2
        module load R/3.6.0
        python3 /u/project/gandalm/cindywen/isoform_twas/sqtl_new/fastqtl-gtex/python/merge_chunks.py \
            {input[0]} \
            {input[1]} \
            gtex \
            --fdr 0.05 \
            -o {params.outdir} \
        """


rule make_torus_input:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/gtex.allpairs.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/gtex.allpairs.torus.txt.gz",
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=480,
    shell:
        """
        zcat {input[0]} | tail -n+2 | awk '{{print $1,$2,$3,$7,$8,$9}}' OFS=' ' | gzip > {output[0]}
        """


# Celine ran this in bash
rule run_torus:
    input:
        annot="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/annot/mixed_variant_annot_ensembl_vep.txt.gz",
        fastqtl="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/gtex.allpairs.torus.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/torus.out",
    resources:
        mem_gb=32,
        time_min=4800,
        num_cores=16,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load python/3.7.2

        /u/project/gandalm/shared/apps/torus-master/src/torus \
            -d {input.fastqtl} \
            --fastqtl \
            -est \
            -annot {input.annot} \
            > {output[0]}
        """


################################### qvalue pi0 ###################################
# e/isoQTL scripts in eqtl_new/code
# here process sQTL large files
# first generated all.introns.tested.tsv, "NA" for unmapped gene
rule qvalue_pi0_prep_chunk:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/chunk{chunk}.txt.gz",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/chunk{chunk}.txt.gz.done",
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/leafviz_annot/all.introns.tested.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/prep.{chunk}.txt.gz",
    resources:
        mem_gb=4,
        time_min=120,
    shell:
        """
        mkdir -p /u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/

        # 1. merge with intron annotation
        # 2. merge gene_qtl
        # 3. only keep lines containing/starting with "ENSG", remove NA gene entries
        # 4. sort by gene_qtl and npval, then remove duplicate gene_qtl, keep the lowest npval pair (sort -k1 is sorting by both gene_qtl and npval columns)

        join -j 1 -o 2.2,1.2,1.4 <(sort -k1 <(zcat {input[0]})) <(sort -k1 {input[2]}) | \
        awk '{{print $1":"$2" "$3}}' | \
        awk '$1 ~ /ENSG/ {{ print }}' | \
        sort -k1,1 -k2,2n | \
        awk '!seen[$1]++' | \
        gzip > {output[0]}
        """


rule qvalue_pi0_prep_merge:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/prep.{chunk}.txt.gz",
            chunk=np.arange(1, 101, 1),
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/prep.txt.gz",
    resources:
        mem_gb=4,
        time_min=120,
    params:
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/",
    shell:
        """ 
        zcat {params.outdir}prep.*.txt.gz | \
        sort -k1,1 -k2,2n | \
        awk '!seen[$1]++' | \
        gzip -c > {params.outdir}prep.txt.gz
        """


rule qvalue_pi0:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/prep.txt.gz",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/mixed_nominal_40hcp_1e6/qvalue_pi0_prep/qvalue.done",
    resources:
        mem_gb=4,
        time_min=120,
        num_cores=4,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript scripts/qvalue_pi0_s.R
        touch {output[0]}
        """


################################### trimester-specific ###################################
rule tri_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/lc.tri{tri}.bed.gz",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri_specific/tri{tri}_{num_hcp}HCP_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        num_gpc=5,
        outdir=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri_specific/"
        ),
        prefix="tri{tri}",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/decon_cov.R \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir} \
            --prefix {params.prefix}
        """


rule vcf_to_dosage_trimester:
    input:
        vcf="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/filtered.hg19.sorted.removeRel.vcf.gz",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/{file}",
            file=[
                "4_columns.tsv",
                "sample_list.tsv",
                "header.tsv",
                "header_row.tsv.gz",
                "dose_matrix.tsv.gz",
                "filtered.hg19.sorted.removeRel.dose.tsv.gz",
                "filtered.hg19.sorted.removeRel.dose.tsv.gz.tbi",
            ],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/filtered.hg19.sorted.removeRel",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load bcftools/1.11
        module load htslib/1.12

        # Extract header
        printf 'CHROM\nPOS\nREF\nALT\n' > {output[0]}
        bcftools query -l {input.vcf} > {output[1]}
        cat {output[0]} {output[1]} > {output[2]}
        {config[CSVTK]} transpose {output[2]} -T | gzip > {output[3]}

        # Extract dosage and merge
        bcftools +dosage {input.vcf} -- -t GT | tail -n+2 | gzip > {output[4]}
        zcat {output[3]} {output[4]} | bgzip > {params.prefix}.dose.tsv.gz
        tabix -s1 -b2 -e2 -S1 {params.prefix}.dose.tsv.gz
        """


rule make_susie_meta_trimester:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/lc.tri{tri}.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri{tri}_sample_meta_{num_hcp}hcp.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    params:
        qtl_group="tri{tri}_cissqtl_{num_hcp}hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_meta.R \
            --bed {input[0]} \
            --qtl_group {params.qtl_group} \
            --exclude_rel {input[1]}
        """


rule make_susie_expr_trimester:
    input:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/lc.tri{tri}.bed.gz",
        "/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/related.txt",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri{tri}.susie.tsv",
    resources:
        mem_gb=4,
        time_min=60,
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/make_susie_expr.R \
            --input {input[0]} \
            --outname {output[0]} \
            --exclude_rel {input[1]}
        """


rule run_susie_trimester:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri{tri}.susie.tsv",
        pheno_meta="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/cluster/phenotype_meta.tsv",
        sample_meta="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri{tri}_sample_meta_{num_hcp}hcp.tsv",
        sig_pheno="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/tri{tri}_perm_{num_hcp}hcp/sig_pheno.txt",
        genotype_matrix="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/eur/filtered.hg19.sorted.removeRel.dose.tsv.gz",
        covariates="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/eur/tri_specific/tri{tri}_{num_hcp}HCP_cov.txt",
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{{tri}}_cissqtl_{{num_hcp}}hcp_perm.{{batch}}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
        ),
    resources:
        mem_gb=4,
        num_cores=4,
        time_min=120,
    params:
        cisdistance=1000000,
        permuted="true",
        qtl_group="tri{tri}_cissqtl_{num_hcp}hcp_perm",
        out_dir=(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/"
        ),
        quant_method="leafcutter",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        mkdir -p {params.out_dir}

        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/run_susie_customized.R \
            --expression_matrix {input.expr} \
            --phenotype_meta {input.pheno_meta} \
            --sample_meta {input.sample_meta} \
            --sig_pheno {input.sig_pheno} \
            --covariates {input.covariates} \
            --genotype_matrix {input.genotype_matrix} \
            --chunk '{wildcards.batch} {config[SUSIE_N_BATCH]}' \
            --cisdistance {params.cisdistance} \
            --out_prefix {params.out_dir}{params.qtl_group}.{wildcards.batch}_{config[SUSIE_N_BATCH]} \
            --permuted {params.permuted} \
            --qtl_group {params.qtl_group} \
            --quant_method {params.quant_method}
        """


rule merge_susie_trimester:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{{tri}}_cissqtl_{{num_hcp}}hcp_perm.{batch}_"
            + config["SUSIE_N_BATCH"]
            + ".{file}",
            file=["txt", "cred.txt", "snp.txt"],
            batch=np.arange(1, 501, 1),
        ),
    output:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{{tri}}_cissqtl_{{num_hcp}}hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz"],
        ),
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{tri}_cissqtl_{num_hcp}hcp_perm",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.12
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.txt | bgzip -c > {params.prefix}.txt.gz
        awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.cred.txt | bgzip -c > {params.prefix}.cred.txt.gz
        # awk 'NR == 1 || FNR > 1{{print}}' {params.prefix}.*_{config[SUSIE_N_BATCH]}.snp.txt | bgzip -c > {params.prefix}.snp.txt.gz
        """


rule sort_susie_trimester:
    input:
        expand(
            "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{{tri}}_cissqtl_{{num_hcp}}hcp_perm.{file}",
            file=["txt.gz", "cred.txt.gz"],
        ),
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{tri}_cissqtl_{num_hcp}hcp_perm_purity_filtered.txt.gz",
    resources:
        mem_gb=6,
        time_min=120,
    params:
        prefix="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/results/susie_finemap/tri{tri}_cissqtl_{num_hcp}hcp_perm",
    shell:
        """ 
        . /u/local/Modules/default/init/modules.sh
        module load htslib/1.12
        gunzip -c {input[0]} > {params.prefix}_susie_merged.txt
        (head -n 1 {params.prefix}_susie_merged.txt && tail -n +2 {params.prefix}_susie_merged.txt | sort -k3 -k4n ) | bgzip > {params.prefix}_purity_filtered.txt.gz
        """


################################### sex-specific ###################################
rule sex_cov:
    input:
        expr="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/lc.{sex}.bed.gz",
        picard=(
            "/u/project/gandalm/cindywen/isoform_twas/picard/picard_QC_compiled.tsv"
        ),
        geno_pc="/u/project/gandalm/cindywen/isoform_twas/genotype/all_data/isec_R2_greater_than_3/ancestry/pca/data.ref.eigenvec",
        meta="/u/project/gandalm/cindywen/isoform_twas/eqtl_new/metadata_inferSex.tsv",
    output:
        "/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/sex_specific/{sex}_{num_hcp}HCP_cov.txt",
    resources:
        mem_gb=8,
        time_min=60,
    params:
        num_gpc=5,
        outdir="/u/project/gandalm/cindywen/isoform_twas/sqtl_new/data/sex_specific/",
        prefix="{sex}",
    shell:
        """
        . /u/local/Modules/default/init/modules.sh
        module load R/4.1.0-BIO
        mkdir -p {params.outdir}
        Rscript /u/project/gandalm/cindywen/isoform_twas/eqtl_new/code/scripts/decon_cov.R \
            --num_hcp {wildcards.num_hcp} \
            --expr {input.expr} \
            --picard {input.picard} \
            --geno_pc {input.geno_pc} \
            --meta {input.meta} \
            --num_gpc {params.num_gpc} \
            --outdir {params.outdir} \
            --prefix {params.prefix}
        """
